// @react-spectrum/form/dist/main.js
require("./main.css");

var _babelRuntimeHelpersExtends = $parcel$interopDefault(require("@babel/runtime/helpers/extends"));

var _babelRuntimeHelpersObjectWithoutProperties = $parcel$interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _babelRuntimeHelpersObjectSpread = $parcel$interopDefault(require("@babel/runtime/helpers/objectSpread2"));

var _temp = require("@react-spectrum/utils");

var classNames = _temp.classNames;
var filterDOMProps = _temp.filterDOMProps;
var useDOMRef = _temp.useDOMRef;
var useStyleProps = _temp.useStyleProps;

var _temp2 = require("@react-spectrum/provider");

var Provider = _temp2.Provider;
var useProviderProps = _temp2.useProviderProps;

var _react2 = require("react");

var _react = $parcel$interopDefault(_react2);

var useContext = _react2.useContext;

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/components/fieldlabel/vars.css
var $ce2ba76b00945bd966a5f95584cdaaee$exports = {};
$ce2ba76b00945bd966a5f95584cdaaee$exports = {
  "spectrum-FieldLabel": "_spectrum-FieldLabel_3fb22",
  "spectrum-FieldLabel--positionSide": "_spectrum-FieldLabel--positionSide_3fb22",
  "spectrum-FieldLabel-requiredIcon": "_spectrum-FieldLabel-requiredIcon_3fb22",
  "spectrum-FieldLabel--alignEnd": "_spectrum-FieldLabel--alignEnd_3fb22",
  "spectrum-Field": "_spectrum-Field_3fb22",
  "spectrum-Field--positionTop": "_spectrum-Field--positionTop_3fb22",
  "spectrum-Field-field": "_spectrum-Field-field_3fb22",
  "spectrum-Field--positionSide": "_spectrum-Field--positionSide_3fb22",
  "spectrum-Form": "_spectrum-Form_3fb22",
  "spectrum-Form--positionSide": "_spectrum-Form--positionSide_3fb22",
  "spectrum-Form--positionTop": "_spectrum-Form--positionTop_3fb22",
  "spectrum-Form-itemLabel": "_spectrum-Form-itemLabel_3fb22",
  "is-disabled": "_is-disabled_3fb22"
};

var $ec21c63dc85db6e3c496ffb42$var$FormContext = _react.createContext({});

function useFormProps(props) {
  var ctx = useContext($ec21c63dc85db6e3c496ffb42$var$FormContext);
  return _babelRuntimeHelpersObjectSpread({}, props, {}, ctx);
}

exports.useFormProps = useFormProps;
var $ce2ba76b00945bd966a5f95584cdaaee$$interop$default = $parcel$interopDefault($ce2ba76b00945bd966a5f95584cdaaee$exports);

function $ec21c63dc85db6e3c496ffb42$var$Form(props, ref) {
  props = useProviderProps(props);

  var _props = props,
      children = _props.children,
      _props$labelPosition = _props.labelPosition,
      labelPosition = _props$labelPosition === void 0 ? 'top' : _props$labelPosition,
      _props$labelAlign = _props.labelAlign,
      labelAlign = _props$labelAlign === void 0 ? 'start' : _props$labelAlign,
      isRequired = _props.isRequired,
      necessityIndicator = _props.necessityIndicator,
      isQuiet = _props.isQuiet,
      isEmphasized = _props.isEmphasized,
      isDisabled = _props.isDisabled,
      isReadOnly = _props.isReadOnly,
      validationState = _props.validationState,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(_props, ["children", "labelPosition", "labelAlign", "isRequired", "necessityIndicator", "isQuiet", "isEmphasized", "isDisabled", "isReadOnly", "validationState"]);

  var _useStyleProps = useStyleProps(otherProps),
      styleProps = _useStyleProps.styleProps;

  var domRef = useDOMRef(ref);
  var ctx = {
    labelPosition: labelPosition,
    labelAlign: labelAlign,
    necessityIndicator: necessityIndicator
  };
  return (/*#__PURE__*/_react.createElement("form", _babelRuntimeHelpersExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef,
      className: classNames($ce2ba76b00945bd966a5f95584cdaaee$$interop$default, 'spectrum-Form', {
        'spectrum-Form--positionSide': labelPosition === 'side',
        'spectrum-Form--positionTop': labelPosition === 'top'
      }, styleProps.className)
    }), /*#__PURE__*/_react.createElement($ec21c63dc85db6e3c496ffb42$var$FormContext.Provider, {
      value: ctx
    }, /*#__PURE__*/_react.createElement(Provider, {
      isQuiet: isQuiet,
      isEmphasized: isEmphasized,
      isDisabled: isDisabled,
      isReadOnly: isReadOnly,
      isRequired: isRequired,
      validationState: validationState
    }, children)))
  );
}

var Form = _react.forwardRef($ec21c63dc85db6e3c496ffb42$var$Form);

exports.Form = Form;
