// @react-spectrum/form/dist/module.js
import "./main.css";
import _babelRuntimeHelpersEsmExtends from "@babel/runtime/helpers/esm/extends";
import _babelRuntimeHelpersEsmObjectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _babelRuntimeHelpersEsmObjectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { classNames, filterDOMProps, useDOMRef, useStyleProps } from "@react-spectrum/utils";
import { Provider, useProviderProps } from "@react-spectrum/provider";
import _react, { useContext } from "react";

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/components/fieldlabel/vars.css
var $c356388cc3519f2d617e9a5ee79a$exports = {};
$c356388cc3519f2d617e9a5ee79a$exports = {
  "spectrum-FieldLabel": "_spectrum-FieldLabel_3fb22",
  "spectrum-FieldLabel--positionSide": "_spectrum-FieldLabel--positionSide_3fb22",
  "spectrum-FieldLabel-requiredIcon": "_spectrum-FieldLabel-requiredIcon_3fb22",
  "spectrum-FieldLabel--alignEnd": "_spectrum-FieldLabel--alignEnd_3fb22",
  "spectrum-Field": "_spectrum-Field_3fb22",
  "spectrum-Field--positionTop": "_spectrum-Field--positionTop_3fb22",
  "spectrum-Field-field": "_spectrum-Field-field_3fb22",
  "spectrum-Field--positionSide": "_spectrum-Field--positionSide_3fb22",
  "spectrum-Form": "_spectrum-Form_3fb22",
  "spectrum-Form--positionSide": "_spectrum-Form--positionSide_3fb22",
  "spectrum-Form--positionTop": "_spectrum-Form--positionTop_3fb22",
  "spectrum-Form-itemLabel": "_spectrum-Form-itemLabel_3fb22",
  "is-disabled": "_is-disabled_3fb22"
};

let $f2742115939d8dea25a9b7ab0f4f0$var$FormContext = _react.createContext({});

export function useFormProps(props) {
  let ctx = useContext($f2742115939d8dea25a9b7ab0f4f0$var$FormContext);
  return _babelRuntimeHelpersEsmObjectSpread({}, props, {}, ctx);
}
var $c356388cc3519f2d617e9a5ee79a$$interop$default = $parcel$interopDefault($c356388cc3519f2d617e9a5ee79a$exports);

function $f2742115939d8dea25a9b7ab0f4f0$var$Form(props, ref) {
  props = useProviderProps(props);

  let {
    children,
    labelPosition = 'top',
    labelAlign = 'start',
    isRequired,
    necessityIndicator,
    isQuiet,
    isEmphasized,
    isDisabled,
    isReadOnly,
    validationState
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["children", "labelPosition", "labelAlign", "isRequired", "necessityIndicator", "isQuiet", "isEmphasized", "isDisabled", "isReadOnly", "validationState"]);

  let {
    styleProps
  } = useStyleProps(otherProps);
  let domRef = useDOMRef(ref);
  let ctx = {
    labelPosition,
    labelAlign,
    necessityIndicator
  };
  return (/*#__PURE__*/_react.createElement("form", _babelRuntimeHelpersEsmExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef,
      className: classNames($c356388cc3519f2d617e9a5ee79a$$interop$default, 'spectrum-Form', {
        'spectrum-Form--positionSide': labelPosition === 'side',
        'spectrum-Form--positionTop': labelPosition === 'top'
      }, styleProps.className)
    }), /*#__PURE__*/_react.createElement($f2742115939d8dea25a9b7ab0f4f0$var$FormContext.Provider, {
      value: ctx
    }, /*#__PURE__*/_react.createElement(Provider, {
      isQuiet: isQuiet,
      isEmphasized: isEmphasized,
      isDisabled: isDisabled,
      isReadOnly: isReadOnly,
      isRequired: isRequired,
      validationState: validationState
    }, children)))
  );
}

export const Form = _react.forwardRef($f2742115939d8dea25a9b7ab0f4f0$var$Form);
