// @react-spectrum/label/dist/main.js
require("./main.css");

var _babelRuntimeHelpersExtends = $parcel$interopDefault(require("@babel/runtime/helpers/extends"));

var _babelRuntimeHelpersObjectWithoutProperties = $parcel$interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _babelRuntimeHelpersInteropRequireDefault = $parcel$interopDefault(require("@babel/runtime/helpers/interopRequireDefault"));

var _spectrumIconsUiAsterisk = $parcel$interopDefault(require("@spectrum-icons/ui/Asterisk"));

var _temp = require("@react-spectrum/utils");

var classNames = _temp.classNames;
var filterDOMProps = _temp.filterDOMProps;
var useDOMRef = _temp.useDOMRef;
var useStyleProps = _temp.useStyleProps;

var _react = $parcel$interopDefault(require("react"));

var useMessageFormatter = require("@react-aria/i18n").useMessageFormatter;

var useProviderProps = require("@react-spectrum/provider").useProviderProps;

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ar-AE.json
var $d071ce38d07d90292b6c3df54e3d9fab$exports = {};
$d071ce38d07d90292b6c3df54e3d9fab$exports = JSON.parse("{\"(optional)\":\"(اختياري)\",\"(required)\":\"(مطلوب)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/bg-BG.json
var $e293d50ff40ef1e299f479abcdf0937f$exports = {};
$e293d50ff40ef1e299f479abcdf0937f$exports = JSON.parse("{\"(optional)\":\"(optional)\",\"(required)\":\"(required)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/cs-CZ.json
var $cd7a3d695f60c229a95564658460233f$exports = {};
$cd7a3d695f60c229a95564658460233f$exports = JSON.parse("{\"(optional)\":\"(volitelně)\",\"(required)\":\"(požadováno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/da-DK.json
var $ba6d04a1e3e1e9be41738211489a$exports = {};
$ba6d04a1e3e1e9be41738211489a$exports = JSON.parse("{\"(optional)\":\"(valgfrit)\",\"(required)\":\"(påkrævet)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/de-DE.json
var $c5f8da166151b638cdb852bd4eedd67$exports = {};
$c5f8da166151b638cdb852bd4eedd67$exports = JSON.parse("{\"(optional)\":\"(optional)\",\"(required)\":\"(erforderlich)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/el-GR.json
var $b8df4ee65138b009af81a6632d37$exports = {};
$b8df4ee65138b009af81a6632d37$exports = JSON.parse("{\"(optional)\":\"(προαιρετικό)\",\"(required)\":\"(απαιτείται)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/en-US.json
var $e01158763064bfb0f4a30318d5571ac4$exports = {};
$e01158763064bfb0f4a30318d5571ac4$exports = JSON.parse("{\"(required)\":\"(required)\",\"(optional)\":\"(optional)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/es-ES.json
var $be525af0885e8288add0055af80fc1$exports = {};
$be525af0885e8288add0055af80fc1$exports = JSON.parse("{\"(optional)\":\"(opcional)\",\"(required)\":\"(necesario)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/et-EE.json
var $f8540c5ccc9de0c18d6ddfb442798d2$exports = {};
$f8540c5ccc9de0c18d6ddfb442798d2$exports = JSON.parse("{\"(optional)\":\"(valikuline)\",\"(required)\":\"(nõutav)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/fi-FI.json
var $ee97bc8cfe6ed0ea55ae30b981c2fe5$exports = {};
$ee97bc8cfe6ed0ea55ae30b981c2fe5$exports = JSON.parse("{\"(optional)\":\"(valinnainen)\",\"(required)\":\"(pakollinen)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/fr-FR.json
var $f5933811e41800eae1f5bea09d6d9497$exports = {};
$f5933811e41800eae1f5bea09d6d9497$exports = JSON.parse("{\"(optional)\":\"(facultatif)\",\"(required)\":\"(requis)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/he-IL.json
var $ae55961b432ff505212d42a383a2dc0a$exports = {};
$ae55961b432ff505212d42a383a2dc0a$exports = JSON.parse("{\"(optional)\":\"(optional)\",\"(required)\":\"(required)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/hr-HR.json
var $cd2210e52a70bd895f4e02b1$exports = {};
$cd2210e52a70bd895f4e02b1$exports = JSON.parse("{\"(optional)\":\"(opcionalno)\",\"(required)\":\"(obvezno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/hu-HU.json
var $ab98b736b2df688226210fff000ec0c0$exports = {};
$ab98b736b2df688226210fff000ec0c0$exports = JSON.parse("{\"(optional)\":\"(opcionális)\",\"(required)\":\"(kötelező)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/it-IT.json
var $ec6ecbe35953fc158f030ce4d9054$exports = {};
$ec6ecbe35953fc158f030ce4d9054$exports = JSON.parse("{\"(optional)\":\"(facoltativo)\",\"(required)\":\"(obbligatorio)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ja-JP.json
var $d6872d0c307203ae85698ca76b330$exports = {};
$d6872d0c307203ae85698ca76b330$exports = JSON.parse("{\"(optional)\":\"（オプション）\",\"(required)\":\"（必須）\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ko-KR.json
var $efd81f5e5622df7089190a95735a8a$exports = {};
$efd81f5e5622df7089190a95735a8a$exports = JSON.parse("{\"(optional)\":\"(선택 사항)\",\"(required)\":\"(필수 사항)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/lt-LT.json
var $f8f581a0870f61ec0ac73be27f1505b$exports = {};
$f8f581a0870f61ec0ac73be27f1505b$exports = JSON.parse("{\"(optional)\":\"(pasirenkama)\",\"(required)\":\"(privaloma)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/lv-LV.json
var $c9c94fded7c6182b9d26b74a72190f$exports = {};
$c9c94fded7c6182b9d26b74a72190f$exports = JSON.parse("{\"(optional)\":\"(neobligāti)\",\"(required)\":\"(obligāti)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/nb-NO.json
var $f914c0353aee02755109f7bb5bb13a$exports = {};
$f914c0353aee02755109f7bb5bb13a$exports = JSON.parse("{\"(optional)\":\"(valgfritt)\",\"(required)\":\"(obligatorisk)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/nl-NL.json
var $dd96646f560fafcb8bdf73bb030d27b5$exports = {};
$dd96646f560fafcb8bdf73bb030d27b5$exports = JSON.parse("{\"(optional)\":\"(optioneel)\",\"(required)\":\"(vereist)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/pl-PL.json
var $b738c997e942a38363e7ed8e3cb68b$exports = {};
$b738c997e942a38363e7ed8e3cb68b$exports = JSON.parse("{\"(optional)\":\"(opcjonalne)\",\"(required)\":\"(wymagane)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/pt-BR.json
var $b9c0867d8c38bb72132bb08b8ca7690$exports = {};
$b9c0867d8c38bb72132bb08b8ca7690$exports = JSON.parse("{\"(optional)\":\"(opcional)\",\"(required)\":\"(obrigatório)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ro-RO.json
var $e3454cfc24d366598f19a73b8d636356$exports = {};
$e3454cfc24d366598f19a73b8d636356$exports = JSON.parse("{\"(optional)\":\"(opţional)\",\"(required)\":\"(obligatoriu)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ru-RU.json
var $c91f305f8bb6c75662c707561913cf$exports = {};
$c91f305f8bb6c75662c707561913cf$exports = JSON.parse("{\"(optional)\":\"(дополнительно)\",\"(required)\":\"(требуется)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sk-SK.json
var $bd34bdf786111560f73fe81dc2403f16$exports = {};
$bd34bdf786111560f73fe81dc2403f16$exports = JSON.parse("{\"(optional)\":\"(nepovinné)\",\"(required)\":\"(povinné)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sl-SI.json
var $fbb8e9d2a7ac5661da97beedd331cfe$exports = {};
$fbb8e9d2a7ac5661da97beedd331cfe$exports = JSON.parse("{\"(optional)\":\"(opcijsko)\",\"(required)\":\"(obvezno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sr-SP.json
var $d73ba941baada77b2cb9b6b9f8ecbea$exports = {};
$d73ba941baada77b2cb9b6b9f8ecbea$exports = JSON.parse("{\"(optional)\":\"(opciono)\",\"(required)\":\"(obavezno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sv-SE.json
var $e562d84960292f7d443819e247a165$exports = {};
$e562d84960292f7d443819e247a165$exports = JSON.parse("{\"(optional)\":\"(valfritt)\",\"(required)\":\"(krävs)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/tr-TR.json
var $faf11825fdfa9cfa96ac2d2b05f398$exports = {};
$faf11825fdfa9cfa96ac2d2b05f398$exports = JSON.parse("{\"(optional)\":\"(isteğe bağlı)\",\"(required)\":\"(gerekli)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/uk-UA.json
var $b61672a313b3dc313e647edc22340a14$exports = {};
$b61672a313b3dc313e647edc22340a14$exports = JSON.parse("{\"(optional)\":\"(необов’язково)\",\"(required)\":\"(обов’язково)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/zh-CN.json
var $d5e1d89fe88b676751f1a16d441be4c3$exports = {};
$d5e1d89fe88b676751f1a16d441be4c3$exports = JSON.parse("{\"(optional)\":\"（可选）\",\"(required)\":\"（必填）\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/zh-TW.json
var $b4d6a1a30d3fa980eb7aee19602055a$exports = {};
$b4d6a1a30d3fa980eb7aee19602055a$exports = JSON.parse("{\"(optional)\":\"(選填)\",\"(required)\":\"(必填)\"}");
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/components/fieldlabel/vars.css
var $ce2ba76b00945bd966a5f95584cdaaee$exports = {};
$ce2ba76b00945bd966a5f95584cdaaee$exports = {
  "spectrum-FieldLabel": "_spectrum-FieldLabel_3fb22",
  "spectrum-FieldLabel--positionSide": "_spectrum-FieldLabel--positionSide_3fb22",
  "spectrum-FieldLabel-requiredIcon": "_spectrum-FieldLabel-requiredIcon_3fb22",
  "spectrum-FieldLabel--alignEnd": "_spectrum-FieldLabel--alignEnd_3fb22",
  "spectrum-Field": "_spectrum-Field_3fb22",
  "spectrum-Field--positionTop": "_spectrum-Field--positionTop_3fb22",
  "spectrum-Field-field": "_spectrum-Field-field_3fb22",
  "spectrum-Field--positionSide": "_spectrum-Field--positionSide_3fb22",
  "spectrum-Form": "_spectrum-Form_3fb22",
  "spectrum-Form--positionSide": "_spectrum-Form--positionSide_3fb22",
  "spectrum-Form--positionTop": "_spectrum-Form--positionTop_3fb22",
  "spectrum-Form-itemLabel": "_spectrum-Form-itemLabel_3fb22",
  "is-disabled": "_is-disabled_3fb22"
};
// @ts-ignore
var $a4d09ce874af58263db10205c6c322d$var$intlMessages = {
  "ar-AE": _babelRuntimeHelpersInteropRequireDefault($d071ce38d07d90292b6c3df54e3d9fab$exports).default,
  "bg-BG": _babelRuntimeHelpersInteropRequireDefault($e293d50ff40ef1e299f479abcdf0937f$exports).default,
  "cs-CZ": _babelRuntimeHelpersInteropRequireDefault($cd7a3d695f60c229a95564658460233f$exports).default,
  "da-DK": _babelRuntimeHelpersInteropRequireDefault($ba6d04a1e3e1e9be41738211489a$exports).default,
  "de-DE": _babelRuntimeHelpersInteropRequireDefault($c5f8da166151b638cdb852bd4eedd67$exports).default,
  "el-GR": _babelRuntimeHelpersInteropRequireDefault($b8df4ee65138b009af81a6632d37$exports).default,
  "en-US": _babelRuntimeHelpersInteropRequireDefault($e01158763064bfb0f4a30318d5571ac4$exports).default,
  "es-ES": _babelRuntimeHelpersInteropRequireDefault($be525af0885e8288add0055af80fc1$exports).default,
  "et-EE": _babelRuntimeHelpersInteropRequireDefault($f8540c5ccc9de0c18d6ddfb442798d2$exports).default,
  "fi-FI": _babelRuntimeHelpersInteropRequireDefault($ee97bc8cfe6ed0ea55ae30b981c2fe5$exports).default,
  "fr-FR": _babelRuntimeHelpersInteropRequireDefault($f5933811e41800eae1f5bea09d6d9497$exports).default,
  "he-IL": _babelRuntimeHelpersInteropRequireDefault($ae55961b432ff505212d42a383a2dc0a$exports).default,
  "hr-HR": _babelRuntimeHelpersInteropRequireDefault($cd2210e52a70bd895f4e02b1$exports).default,
  "hu-HU": _babelRuntimeHelpersInteropRequireDefault($ab98b736b2df688226210fff000ec0c0$exports).default,
  "it-IT": _babelRuntimeHelpersInteropRequireDefault($ec6ecbe35953fc158f030ce4d9054$exports).default,
  "ja-JP": _babelRuntimeHelpersInteropRequireDefault($d6872d0c307203ae85698ca76b330$exports).default,
  "ko-KR": _babelRuntimeHelpersInteropRequireDefault($efd81f5e5622df7089190a95735a8a$exports).default,
  "lt-LT": _babelRuntimeHelpersInteropRequireDefault($f8f581a0870f61ec0ac73be27f1505b$exports).default,
  "lv-LV": _babelRuntimeHelpersInteropRequireDefault($c9c94fded7c6182b9d26b74a72190f$exports).default,
  "nb-NO": _babelRuntimeHelpersInteropRequireDefault($f914c0353aee02755109f7bb5bb13a$exports).default,
  "nl-NL": _babelRuntimeHelpersInteropRequireDefault($dd96646f560fafcb8bdf73bb030d27b5$exports).default,
  "pl-PL": _babelRuntimeHelpersInteropRequireDefault($b738c997e942a38363e7ed8e3cb68b$exports).default,
  "pt-BR": _babelRuntimeHelpersInteropRequireDefault($b9c0867d8c38bb72132bb08b8ca7690$exports).default,
  "ro-RO": _babelRuntimeHelpersInteropRequireDefault($e3454cfc24d366598f19a73b8d636356$exports).default,
  "ru-RU": _babelRuntimeHelpersInteropRequireDefault($c91f305f8bb6c75662c707561913cf$exports).default,
  "sk-SK": _babelRuntimeHelpersInteropRequireDefault($bd34bdf786111560f73fe81dc2403f16$exports).default,
  "sl-SI": _babelRuntimeHelpersInteropRequireDefault($fbb8e9d2a7ac5661da97beedd331cfe$exports).default,
  "sr-SP": _babelRuntimeHelpersInteropRequireDefault($d73ba941baada77b2cb9b6b9f8ecbea$exports).default,
  "sv-SE": _babelRuntimeHelpersInteropRequireDefault($e562d84960292f7d443819e247a165$exports).default,
  "tr-TR": _babelRuntimeHelpersInteropRequireDefault($faf11825fdfa9cfa96ac2d2b05f398$exports).default,
  "uk-UA": _babelRuntimeHelpersInteropRequireDefault($b61672a313b3dc313e647edc22340a14$exports).default,
  "zh-CN": _babelRuntimeHelpersInteropRequireDefault($d5e1d89fe88b676751f1a16d441be4c3$exports).default,
  "zh-TW": _babelRuntimeHelpersInteropRequireDefault($b4d6a1a30d3fa980eb7aee19602055a$exports).default
};
var $ce2ba76b00945bd966a5f95584cdaaee$$interop$default = $parcel$interopDefault($ce2ba76b00945bd966a5f95584cdaaee$exports);

function $a4d09ce874af58263db10205c6c322d$var$Label(props, ref) {
  props = useProviderProps(props);

  var _props = props,
      children = _props.children,
      _props$labelPosition = _props.labelPosition,
      labelPosition = _props$labelPosition === void 0 ? 'top' : _props$labelPosition,
      _props$labelAlign = _props.labelAlign,
      labelAlign = _props$labelAlign === void 0 ? labelPosition === 'side' ? 'start' : null : _props$labelAlign,
      isRequired = _props.isRequired,
      _props$necessityIndic = _props.necessityIndicator,
      necessityIndicator = _props$necessityIndic === void 0 ? isRequired != null ? 'icon' : null : _props$necessityIndic,
      htmlFor = _props.htmlFor,
      labelFor = _props.for,
      _props$elementType = _props.elementType,
      ElementType = _props$elementType === void 0 ? 'label' : _props$elementType,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(_props, ["children", "labelPosition", "labelAlign", "isRequired", "necessityIndicator", "htmlFor", "for", "elementType"]);

  var domRef = useDOMRef(ref);

  var _useStyleProps = useStyleProps(otherProps),
      styleProps = _useStyleProps.styleProps;

  var formatMessage = useMessageFormatter($a4d09ce874af58263db10205c6c322d$var$intlMessages);
  var necessityLabel = isRequired ? formatMessage('(required)') : formatMessage('(optional)');

  var icon = /*#__PURE__*/_react.createElement(_spectrumIconsUiAsterisk, {
    UNSAFE_className: classNames($ce2ba76b00945bd966a5f95584cdaaee$$interop$default, 'spectrum-FieldLabel-requiredIcon'),
    size: "S"
  });

  var labelClassNames = classNames($ce2ba76b00945bd966a5f95584cdaaee$$interop$default, 'spectrum-FieldLabel', {
    'spectrum-FieldLabel--positionSide': labelPosition === 'side',
    'spectrum-FieldLabel--alignEnd': labelAlign === 'end'
  }, styleProps.className);
  return (/*#__PURE__*/_react.createElement(ElementType, _babelRuntimeHelpersExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef,
      className: labelClassNames,
      htmlFor: ElementType === 'label' ? labelFor || htmlFor : undefined
    }), children, necessityIndicator && " \u200B", necessityIndicator === 'label' && /*#__PURE__*/_react.createElement("span", {
      "aria-hidden": isRequired
    }, necessityLabel), necessityIndicator === 'icon' && isRequired && icon)
  );
}

var Label = _react.forwardRef($a4d09ce874af58263db10205c6c322d$var$Label);

exports.Label = Label;
