// @react-spectrum/label/dist/module.js
import "./main.css";
import _babelRuntimeHelpersEsmExtends from "@babel/runtime/helpers/esm/extends";
import _babelRuntimeHelpersEsmObjectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _babelRuntimeHelpersEsmInteropRequireDefault from "@babel/runtime/helpers/esm/interopRequireDefault";
import _spectrumIconsUiAsterisk from "@spectrum-icons/ui/Asterisk";
import { classNames, filterDOMProps, useDOMRef, useStyleProps } from "@react-spectrum/utils";
import _react from "react";
import { useMessageFormatter } from "@react-aria/i18n";
import { useProviderProps } from "@react-spectrum/provider";

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ar-AE.json
var $a63c56191af40ea22f66b7c61776fde$exports = {};
$a63c56191af40ea22f66b7c61776fde$exports = JSON.parse("{\"(optional)\":\"(اختياري)\",\"(required)\":\"(مطلوب)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/bg-BG.json
var $b921fe9ae8f83b72ab9684a10db9a38c$exports = {};
$b921fe9ae8f83b72ab9684a10db9a38c$exports = JSON.parse("{\"(optional)\":\"(optional)\",\"(required)\":\"(required)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/cs-CZ.json
var $b52f74c039e548c0a5a6daa22fad312$exports = {};
$b52f74c039e548c0a5a6daa22fad312$exports = JSON.parse("{\"(optional)\":\"(volitelně)\",\"(required)\":\"(požadováno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/da-DK.json
var $a943da95cee671fee2b4a6040e9c7693$exports = {};
$a943da95cee671fee2b4a6040e9c7693$exports = JSON.parse("{\"(optional)\":\"(valgfrit)\",\"(required)\":\"(påkrævet)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/de-DE.json
var $d1a5f951cfcc838cae759581da10551$exports = {};
$d1a5f951cfcc838cae759581da10551$exports = JSON.parse("{\"(optional)\":\"(optional)\",\"(required)\":\"(erforderlich)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/el-GR.json
var $db5c55a258af477f883f4a74a6b9d0c$exports = {};
$db5c55a258af477f883f4a74a6b9d0c$exports = JSON.parse("{\"(optional)\":\"(προαιρετικό)\",\"(required)\":\"(απαιτείται)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/en-US.json
var $a7a4c014289a043c15d3efe2b1755747$exports = {};
$a7a4c014289a043c15d3efe2b1755747$exports = JSON.parse("{\"(required)\":\"(required)\",\"(optional)\":\"(optional)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/es-ES.json
var $fd49644c52581f0be49dcf282d2034f$exports = {};
$fd49644c52581f0be49dcf282d2034f$exports = JSON.parse("{\"(optional)\":\"(opcional)\",\"(required)\":\"(necesario)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/et-EE.json
var $b483f718a475614752b244e50bd7ab76$exports = {};
$b483f718a475614752b244e50bd7ab76$exports = JSON.parse("{\"(optional)\":\"(valikuline)\",\"(required)\":\"(nõutav)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/fi-FI.json
var $d70e5de6297d6896d8d8287f7eb92a9$exports = {};
$d70e5de6297d6896d8d8287f7eb92a9$exports = JSON.parse("{\"(optional)\":\"(valinnainen)\",\"(required)\":\"(pakollinen)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/fr-FR.json
var $fda552b8fd06cfc49e2cde38e5a2e$exports = {};
$fda552b8fd06cfc49e2cde38e5a2e$exports = JSON.parse("{\"(optional)\":\"(facultatif)\",\"(required)\":\"(requis)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/he-IL.json
var $f48f8c47d7d0473d5c7da24c220c4daf$exports = {};
$f48f8c47d7d0473d5c7da24c220c4daf$exports = JSON.parse("{\"(optional)\":\"(optional)\",\"(required)\":\"(required)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/hr-HR.json
var $cef78e0fe0c27084f8cbe598032f3$exports = {};
$cef78e0fe0c27084f8cbe598032f3$exports = JSON.parse("{\"(optional)\":\"(opcionalno)\",\"(required)\":\"(obvezno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/hu-HU.json
var $be6d0303e0d67fcfa7c9af863d1881c0$exports = {};
$be6d0303e0d67fcfa7c9af863d1881c0$exports = JSON.parse("{\"(optional)\":\"(opcionális)\",\"(required)\":\"(kötelező)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/it-IT.json
var $d23b14627b40c21bff8c20e33cc27$exports = {};
$d23b14627b40c21bff8c20e33cc27$exports = JSON.parse("{\"(optional)\":\"(facoltativo)\",\"(required)\":\"(obbligatorio)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ja-JP.json
var $a045c465443a5720c96b73dc63fd8cec$exports = {};
$a045c465443a5720c96b73dc63fd8cec$exports = JSON.parse("{\"(optional)\":\"（オプション）\",\"(required)\":\"（必須）\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ko-KR.json
var $e58204c09e0f2995e8cf1537fe3aab$exports = {};
$e58204c09e0f2995e8cf1537fe3aab$exports = JSON.parse("{\"(optional)\":\"(선택 사항)\",\"(required)\":\"(필수 사항)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/lt-LT.json
var $de08c09c72e02d72aba0b8da04551d22$exports = {};
$de08c09c72e02d72aba0b8da04551d22$exports = JSON.parse("{\"(optional)\":\"(pasirenkama)\",\"(required)\":\"(privaloma)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/lv-LV.json
var $d6b0c1cac26d452ee016e22079368$exports = {};
$d6b0c1cac26d452ee016e22079368$exports = JSON.parse("{\"(optional)\":\"(neobligāti)\",\"(required)\":\"(obligāti)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/nb-NO.json
var $ee1a8157d8e373d338859d611439ba86$exports = {};
$ee1a8157d8e373d338859d611439ba86$exports = JSON.parse("{\"(optional)\":\"(valgfritt)\",\"(required)\":\"(obligatorisk)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/nl-NL.json
var $eab9a3154f37f106d0f6cc27efe$exports = {};
$eab9a3154f37f106d0f6cc27efe$exports = JSON.parse("{\"(optional)\":\"(optioneel)\",\"(required)\":\"(vereist)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/pl-PL.json
var $c5ad1bc93b95e7f7379661bfaa$exports = {};
$c5ad1bc93b95e7f7379661bfaa$exports = JSON.parse("{\"(optional)\":\"(opcjonalne)\",\"(required)\":\"(wymagane)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/pt-BR.json
var $aa89a1c53c07a7e3155d24c14a457a82$exports = {};
$aa89a1c53c07a7e3155d24c14a457a82$exports = JSON.parse("{\"(optional)\":\"(opcional)\",\"(required)\":\"(obrigatório)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ro-RO.json
var $f0e5fc684d8b6a7be04667fa420892$exports = {};
$f0e5fc684d8b6a7be04667fa420892$exports = JSON.parse("{\"(optional)\":\"(opţional)\",\"(required)\":\"(obligatoriu)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/ru-RU.json
var $d5e0dbd3458ffd36e88901f4665fbae8$exports = {};
$d5e0dbd3458ffd36e88901f4665fbae8$exports = JSON.parse("{\"(optional)\":\"(дополнительно)\",\"(required)\":\"(требуется)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sk-SK.json
var $fed67ed01e07634400f095cb6ab788e$exports = {};
$fed67ed01e07634400f095cb6ab788e$exports = JSON.parse("{\"(optional)\":\"(nepovinné)\",\"(required)\":\"(povinné)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sl-SI.json
var $a92d0fb877f53e62f5f4c1e5e8eae30$exports = {};
$a92d0fb877f53e62f5f4c1e5e8eae30$exports = JSON.parse("{\"(optional)\":\"(opcijsko)\",\"(required)\":\"(obvezno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sr-SP.json
var $ed108ce1b50086e994b9f22130ab$exports = {};
$ed108ce1b50086e994b9f22130ab$exports = JSON.parse("{\"(optional)\":\"(opciono)\",\"(required)\":\"(obavezno)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/sv-SE.json
var $fe3bb20a68f5fe1f2edf119abdfcc86$exports = {};
$fe3bb20a68f5fe1f2edf119abdfcc86$exports = JSON.parse("{\"(optional)\":\"(valfritt)\",\"(required)\":\"(krävs)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/tr-TR.json
var $cf95cd52e336f0c91967564f382b72$exports = {};
$cf95cd52e336f0c91967564f382b72$exports = JSON.parse("{\"(optional)\":\"(isteğe bağlı)\",\"(required)\":\"(gerekli)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/uk-UA.json
var $ba33dd046f9d5ac64ace44911ef53a4$exports = {};
$ba33dd046f9d5ac64ace44911ef53a4$exports = JSON.parse("{\"(optional)\":\"(необов’язково)\",\"(required)\":\"(обов’язково)\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/zh-CN.json
var $fc31b3e7feae4a22613279e0bbb60678$exports = {};
$fc31b3e7feae4a22613279e0bbb60678$exports = JSON.parse("{\"(optional)\":\"（可选）\",\"(required)\":\"（必填）\"}");
// ASSET: /Users/govett/dev/react-spectrum/packages/@react-spectrum/label/intl/zh-TW.json
var $e09e63d8eeaac452b9c8da0980cf3fdb$exports = {};
$e09e63d8eeaac452b9c8da0980cf3fdb$exports = JSON.parse("{\"(optional)\":\"(選填)\",\"(required)\":\"(必填)\"}");
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/components/fieldlabel/vars.css
var $c356388cc3519f2d617e9a5ee79a$exports = {};
$c356388cc3519f2d617e9a5ee79a$exports = {
  "spectrum-FieldLabel": "_spectrum-FieldLabel_3fb22",
  "spectrum-FieldLabel--positionSide": "_spectrum-FieldLabel--positionSide_3fb22",
  "spectrum-FieldLabel-requiredIcon": "_spectrum-FieldLabel-requiredIcon_3fb22",
  "spectrum-FieldLabel--alignEnd": "_spectrum-FieldLabel--alignEnd_3fb22",
  "spectrum-Field": "_spectrum-Field_3fb22",
  "spectrum-Field--positionTop": "_spectrum-Field--positionTop_3fb22",
  "spectrum-Field-field": "_spectrum-Field-field_3fb22",
  "spectrum-Field--positionSide": "_spectrum-Field--positionSide_3fb22",
  "spectrum-Form": "_spectrum-Form_3fb22",
  "spectrum-Form--positionSide": "_spectrum-Form--positionSide_3fb22",
  "spectrum-Form--positionTop": "_spectrum-Form--positionTop_3fb22",
  "spectrum-Form-itemLabel": "_spectrum-Form-itemLabel_3fb22",
  "is-disabled": "_is-disabled_3fb22"
};
// @ts-ignore
const $beb551b9ed5ad7940b3cac804bf172c6$var$intlMessages = {
  "ar-AE": _babelRuntimeHelpersEsmInteropRequireDefault($a63c56191af40ea22f66b7c61776fde$exports).default,
  "bg-BG": _babelRuntimeHelpersEsmInteropRequireDefault($b921fe9ae8f83b72ab9684a10db9a38c$exports).default,
  "cs-CZ": _babelRuntimeHelpersEsmInteropRequireDefault($b52f74c039e548c0a5a6daa22fad312$exports).default,
  "da-DK": _babelRuntimeHelpersEsmInteropRequireDefault($a943da95cee671fee2b4a6040e9c7693$exports).default,
  "de-DE": _babelRuntimeHelpersEsmInteropRequireDefault($d1a5f951cfcc838cae759581da10551$exports).default,
  "el-GR": _babelRuntimeHelpersEsmInteropRequireDefault($db5c55a258af477f883f4a74a6b9d0c$exports).default,
  "en-US": _babelRuntimeHelpersEsmInteropRequireDefault($a7a4c014289a043c15d3efe2b1755747$exports).default,
  "es-ES": _babelRuntimeHelpersEsmInteropRequireDefault($fd49644c52581f0be49dcf282d2034f$exports).default,
  "et-EE": _babelRuntimeHelpersEsmInteropRequireDefault($b483f718a475614752b244e50bd7ab76$exports).default,
  "fi-FI": _babelRuntimeHelpersEsmInteropRequireDefault($d70e5de6297d6896d8d8287f7eb92a9$exports).default,
  "fr-FR": _babelRuntimeHelpersEsmInteropRequireDefault($fda552b8fd06cfc49e2cde38e5a2e$exports).default,
  "he-IL": _babelRuntimeHelpersEsmInteropRequireDefault($f48f8c47d7d0473d5c7da24c220c4daf$exports).default,
  "hr-HR": _babelRuntimeHelpersEsmInteropRequireDefault($cef78e0fe0c27084f8cbe598032f3$exports).default,
  "hu-HU": _babelRuntimeHelpersEsmInteropRequireDefault($be6d0303e0d67fcfa7c9af863d1881c0$exports).default,
  "it-IT": _babelRuntimeHelpersEsmInteropRequireDefault($d23b14627b40c21bff8c20e33cc27$exports).default,
  "ja-JP": _babelRuntimeHelpersEsmInteropRequireDefault($a045c465443a5720c96b73dc63fd8cec$exports).default,
  "ko-KR": _babelRuntimeHelpersEsmInteropRequireDefault($e58204c09e0f2995e8cf1537fe3aab$exports).default,
  "lt-LT": _babelRuntimeHelpersEsmInteropRequireDefault($de08c09c72e02d72aba0b8da04551d22$exports).default,
  "lv-LV": _babelRuntimeHelpersEsmInteropRequireDefault($d6b0c1cac26d452ee016e22079368$exports).default,
  "nb-NO": _babelRuntimeHelpersEsmInteropRequireDefault($ee1a8157d8e373d338859d611439ba86$exports).default,
  "nl-NL": _babelRuntimeHelpersEsmInteropRequireDefault($eab9a3154f37f106d0f6cc27efe$exports).default,
  "pl-PL": _babelRuntimeHelpersEsmInteropRequireDefault($c5ad1bc93b95e7f7379661bfaa$exports).default,
  "pt-BR": _babelRuntimeHelpersEsmInteropRequireDefault($aa89a1c53c07a7e3155d24c14a457a82$exports).default,
  "ro-RO": _babelRuntimeHelpersEsmInteropRequireDefault($f0e5fc684d8b6a7be04667fa420892$exports).default,
  "ru-RU": _babelRuntimeHelpersEsmInteropRequireDefault($d5e0dbd3458ffd36e88901f4665fbae8$exports).default,
  "sk-SK": _babelRuntimeHelpersEsmInteropRequireDefault($fed67ed01e07634400f095cb6ab788e$exports).default,
  "sl-SI": _babelRuntimeHelpersEsmInteropRequireDefault($a92d0fb877f53e62f5f4c1e5e8eae30$exports).default,
  "sr-SP": _babelRuntimeHelpersEsmInteropRequireDefault($ed108ce1b50086e994b9f22130ab$exports).default,
  "sv-SE": _babelRuntimeHelpersEsmInteropRequireDefault($fe3bb20a68f5fe1f2edf119abdfcc86$exports).default,
  "tr-TR": _babelRuntimeHelpersEsmInteropRequireDefault($cf95cd52e336f0c91967564f382b72$exports).default,
  "uk-UA": _babelRuntimeHelpersEsmInteropRequireDefault($ba33dd046f9d5ac64ace44911ef53a4$exports).default,
  "zh-CN": _babelRuntimeHelpersEsmInteropRequireDefault($fc31b3e7feae4a22613279e0bbb60678$exports).default,
  "zh-TW": _babelRuntimeHelpersEsmInteropRequireDefault($e09e63d8eeaac452b9c8da0980cf3fdb$exports).default
};
var $c356388cc3519f2d617e9a5ee79a$$interop$default = $parcel$interopDefault($c356388cc3519f2d617e9a5ee79a$exports);

function $beb551b9ed5ad7940b3cac804bf172c6$var$Label(props, ref) {
  props = useProviderProps(props);

  let {
    children,
    labelPosition = 'top',
    labelAlign = labelPosition === 'side' ? 'start' : null,
    isRequired,
    necessityIndicator = isRequired != null ? 'icon' : null,
    htmlFor,
    for: labelFor,
    elementType: ElementType = 'label'
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["children", "labelPosition", "labelAlign", "isRequired", "necessityIndicator", "htmlFor", "for", "elementType"]);

  let domRef = useDOMRef(ref);
  let {
    styleProps
  } = useStyleProps(otherProps);
  let formatMessage = useMessageFormatter($beb551b9ed5ad7940b3cac804bf172c6$var$intlMessages);
  let necessityLabel = isRequired ? formatMessage('(required)') : formatMessage('(optional)');

  let icon = /*#__PURE__*/_react.createElement(_spectrumIconsUiAsterisk, {
    UNSAFE_className: classNames($c356388cc3519f2d617e9a5ee79a$$interop$default, 'spectrum-FieldLabel-requiredIcon'),
    size: "S"
  });

  let labelClassNames = classNames($c356388cc3519f2d617e9a5ee79a$$interop$default, 'spectrum-FieldLabel', {
    'spectrum-FieldLabel--positionSide': labelPosition === 'side',
    'spectrum-FieldLabel--alignEnd': labelAlign === 'end'
  }, styleProps.className);
  return (/*#__PURE__*/_react.createElement(ElementType, _babelRuntimeHelpersEsmExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef,
      className: labelClassNames,
      htmlFor: ElementType === 'label' ? labelFor || htmlFor : undefined
    }), children, necessityIndicator && ' \u200b', necessityIndicator === 'label' && /*#__PURE__*/_react.createElement("span", {
      "aria-hidden": isRequired
    }, necessityLabel), necessityIndicator === 'icon' && isRequired && icon)
  );
}

export let Label = _react.forwardRef($beb551b9ed5ad7940b3cac804bf172c6$var$Label);
