// @react-spectrum/utils/dist/main.js
var _babelRuntimeHelpersTypeof = $parcel$interopDefault(require("@babel/runtime/helpers/typeof"));

var _classnames = $parcel$interopDefault(require("classnames"));

var _react2 = require("react");

var _react = $parcel$interopDefault(_react2);

var useEffect = _react2.useEffect;
var useState = _react2.useState;
var useImperativeHandle = _react2.useImperativeHandle;
var useRef = _react2.useRef;
var useContext = _react2.useContext;
var useMemo = _react2.useMemo;
var useLayoutEffect = _react2.useLayoutEffect;

var _babelRuntimeHelpersObjectSpread = $parcel$interopDefault(require("@babel/runtime/helpers/objectSpread2"));

var _babelRuntimeHelpersSlicedToArray = $parcel$interopDefault(require("@babel/runtime/helpers/slicedToArray"));

var _babelRuntimeHelpersObjectWithoutProperties = $parcel$interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var useLocale = require("@react-aria/i18n").useLocale;

var _babelRuntimeHelpersDefineProperty = $parcel$interopDefault(require("@babel/runtime/helpers/defineProperty"));

var mergeProps = require("@react-aria/utils").mergeProps;

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

var shouldKeepSpectrumClassNames = false;
exports.shouldKeepSpectrumClassNames = shouldKeepSpectrumClassNames;

function keepSpectrumClassNames() {
  shouldKeepSpectrumClassNames = true;
  exports.shouldKeepSpectrumClassNames = shouldKeepSpectrumClassNames;
  console.warn('Legacy spectrum-prefixed class names enabled for backward compatibility. ' + 'We recommend replacing instances of CSS overrides targeting spectrum selectors ' + 'in your app with custom class names of your own, and disabling this flag.');
}

exports.keepSpectrumClassNames = keepSpectrumClassNames;

function classNames(cssModule) {
  var classes = [];

  for (var _len = arguments.length, values = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    values[_key - 1] = arguments[_key];
  }

  for (var _i = 0, _values = values; _i < _values.length; _i++) {
    var value = _values[_i];

    if (_babelRuntimeHelpersTypeof(value) === 'object' && value) {
      var mapped = {};

      for (var _key2 in value) {
        if (cssModule[_key2]) {
          mapped[cssModule[_key2]] = value[_key2];
        }

        if (shouldKeepSpectrumClassNames || !cssModule[_key2]) {
          mapped[_key2] = value[_key2];
        }
      }

      classes.push(mapped);
    } else if (typeof value === 'string') {
      if (cssModule[value]) {
        classes.push(cssModule[value]);
      }

      if (shouldKeepSpectrumClassNames || !cssModule[value]) {
        classes.push(value);
      }
    } else {
      classes.push(value);
    }
  }

  return _classnames.apply(void 0, classes);
}

exports.classNames = classNames;

function getWrappedElement(children) {
  var element;

  if (typeof children === 'string') {
    element = /*#__PURE__*/_react.createElement("span", null, children);
  } else {
    element = _react.Children.only(children);
  }

  return element;
}

exports.getWrappedElement = getWrappedElement;

/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
// IMPORTANT: this needs to be synced with the TypeScript definition of DOMProps
var $d7053fa347a72b28c9e$var$DOMPropNames = {
  id: 1,
  tabIndex: 1,
  role: 1,
  'aria-label': 1,
  'aria-labelledby': 1,
  'aria-describedby': 1,
  'aria-controls': 1,
  'aria-owns': 1,
  'aria-hidden': 1
}; // Sync with TypeScript definition of TextInputDOMProps

var TextInputDOMPropNames = {
  autoComplete: 1,
  maxLength: 1,
  minLength: 1,
  name: 1,
  pattern: 1,
  placeholder: 1,
  type: 1,
  inputMode: 1,
  // Clipboard events
  onCopy: 1,
  onCut: 1,
  onPaste: 1,
  // Composition events
  onCompositionEnd: 1,
  onCompositionStart: 1,
  onCompositionUpdate: 1,
  // Selection events
  onSelect: 1,
  // Input events
  onBeforeInput: 1,
  onInput: 1
}; // Based on DOMAttributes from react types

exports.TextInputDOMPropNames = TextInputDOMPropNames;
var DOMEventPropNames = {
  // Focus events
  onFocus: 1,
  onBlur: 1,
  // Keyboard events
  onKeyDown: 1,
  onKeyPress: 1,
  onKeyUp: 1,
  // Mouse events
  onAuxClick: 1,
  onClick: 1,
  onContextMenu: 1,
  onDoubleClick: 1,
  onDrag: 1,
  onDragEnd: 1,
  onDragEnter: 1,
  onDragExit: 1,
  onDragLeave: 1,
  onDragOver: 1,
  onDragStart: 1,
  onDrop: 1,
  onMouseDown: 1,
  onMouseEnter: 1,
  onMouseLeave: 1,
  onMouseMove: 1,
  onMouseOut: 1,
  onMouseOver: 1,
  onMouseUp: 1,
  // Touch events
  onTouchCancel: 1,
  onTouchEnd: 1,
  onTouchMove: 1,
  onTouchStart: 1,
  // Pointer events
  onPointerDown: 1,
  onPointerMove: 1,
  onPointerUp: 1,
  onPointerCancel: 1,
  onPointerEnter: 1,
  onPointerLeave: 1,
  onPointerOver: 1,
  onPointerOut: 1,
  onGotPointerCapture: 1,
  onLostPointerCapture: 1,
  // UI events
  onScroll: 1,
  // Wheel events
  onWheel: 1
};
/**
 * Checking for data-* props
 */

exports.DOMEventPropNames = DOMEventPropNames;
var $d7053fa347a72b28c9e$var$propRe = /^(data-.*)$/; // Filters out all props that aren't valid DOM props or are user defined via override prop obj.

function filterDOMProps(props) {
  var override = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var filterProps = {};

  var propFilter = _babelRuntimeHelpersObjectSpread({}, $d7053fa347a72b28c9e$var$DOMPropNames, {}, override);

  for (var prop in props) {
    if (Object.prototype.hasOwnProperty.call(props, prop) && (propFilter[prop] || $d7053fa347a72b28c9e$var$propRe.test(prop))) {
      filterProps[prop] = props[prop];
    }
  }

  return filterProps;
}

exports.filterDOMProps = filterDOMProps;

function useMediaQuery(query) {
  var supportsMatchMedia = typeof window !== 'undefined' && typeof window.matchMedia === 'function';

  var _useState = useState(function () {
    return supportsMatchMedia ? window.matchMedia(query).matches : false;
  }),
      _useState2 = _babelRuntimeHelpersSlicedToArray(_useState, 2),
      matches = _useState2[0],
      setMatches = _useState2[1];

  useEffect(function () {
    if (!supportsMatchMedia) {
      return;
    }

    var mq = window.matchMedia(query);

    var onChange = function onChange(evt) {
      setMatches(evt.matches);
    };

    mq.addListener(onChange);
    return function () {
      mq.removeListener(onChange);
    };
  }, [supportsMatchMedia, query]);
  return matches;
}

exports.useMediaQuery = useMediaQuery;

function createDOMRef(ref) {
  return {
    UNSAFE_getDOMNode: function UNSAFE_getDOMNode() {
      return ref.current;
    }
  };
}

exports.createDOMRef = createDOMRef;

function createFocusableRef(domRef) {
  var focusableRef = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : domRef;
  return _babelRuntimeHelpersObjectSpread({}, createDOMRef(domRef), {
    focus: function focus() {
      if (focusableRef.current) {
        focusableRef.current.focus();
      }
    }
  });
}

exports.createFocusableRef = createFocusableRef;

function useDOMRef(ref) {
  var domRef = useRef(null);
  useImperativeHandle(ref, function () {
    return createDOMRef(domRef);
  });
  return domRef;
}

exports.useDOMRef = useDOMRef;

function useFocusableRef(ref, focusableRef) {
  var domRef = useRef(null);
  useImperativeHandle(ref, function () {
    return createFocusableRef(domRef, focusableRef);
  });
  return domRef;
}

exports.useFocusableRef = useFocusableRef;

function unwrapDOMRef(ref) {
  return {
    get current() {
      return ref.current && ref.current.UNSAFE_getDOMNode();
    }

  };
}

exports.unwrapDOMRef = unwrapDOMRef;

function $a739655128e17f93c5a267e67462b806$var$_createForOfIteratorHelper(o) {
  if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) {
    if (Array.isArray(o) || (o = $a739655128e17f93c5a267e67462b806$var$_unsupportedIterableToArray(o))) {
      var i = 0;

      var F = function F() {};

      return {
        s: F,
        n: function n() {
          if (i >= o.length) return {
            done: true
          };
          return {
            done: false,
            value: o[i++]
          };
        },
        e: function e(_e) {
          throw _e;
        },
        f: F
      };
    }

    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  var it,
      normalCompletion = true,
      didErr = false,
      err;
  return {
    s: function s() {
      it = o[Symbol.iterator]();
    },
    n: function n() {
      var step = it.next();
      normalCompletion = step.done;
      return step;
    },
    e: function e(_e2) {
      didErr = true;
      err = _e2;
    },
    f: function f() {
      try {
        if (!normalCompletion && it.return != null) it.return();
      } finally {
        if (didErr) throw err;
      }
    }
  };
}

function $a739655128e17f93c5a267e67462b806$var$_unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return $a739655128e17f93c5a267e67462b806$var$_arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(n);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return $a739655128e17f93c5a267e67462b806$var$_arrayLikeToArray(o, minLen);
}

function $a739655128e17f93c5a267e67462b806$var$_arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }

  return arr2;
}

var baseStyleProps = {
  margin: ['margin', dimensionValue],
  marginStart: [$a739655128e17f93c5a267e67462b806$var$rtl('marginLeft', 'marginRight'), dimensionValue],
  marginEnd: [$a739655128e17f93c5a267e67462b806$var$rtl('marginRight', 'marginLeft'), dimensionValue],
  marginLeft: ['marginLeft', dimensionValue],
  marginRight: ['marginRight', dimensionValue],
  marginTop: ['marginTop', dimensionValue],
  marginBottom: ['marginBottom', dimensionValue],
  marginX: [['marginLeft', 'marginRight'], dimensionValue],
  marginY: [['marginTop', 'marginBottom'], dimensionValue],
  width: ['width', dimensionValue],
  height: ['height', dimensionValue],
  minWidth: ['minWidth', dimensionValue],
  minHeight: ['minHeight', dimensionValue],
  maxWidth: ['maxWidth', dimensionValue],
  maxHeight: ['maxHeight', dimensionValue],
  isHidden: ['display', $a739655128e17f93c5a267e67462b806$var$hiddenValue],
  alignSelf: ['alignSelf', passthroughStyle],
  justifySelf: ['justifySelf', passthroughStyle],
  position: ['position', $a739655128e17f93c5a267e67462b806$var$anyValue],
  zIndex: ['zIndex', $a739655128e17f93c5a267e67462b806$var$anyValue],
  top: ['top', dimensionValue],
  bottom: ['bottom', dimensionValue],
  start: [$a739655128e17f93c5a267e67462b806$var$rtl('left', 'right'), dimensionValue],
  end: [$a739655128e17f93c5a267e67462b806$var$rtl('right', 'left'), dimensionValue],
  left: ['left', dimensionValue],
  right: ['right', dimensionValue],
  flex: ['flex', passthroughStyle],
  flexGrow: ['flexGrow', passthroughStyle],
  flexShrink: ['flexShrink', passthroughStyle],
  flexBasis: ['flexBasis', passthroughStyle]
};
exports.baseStyleProps = baseStyleProps;

var viewStyleProps = _babelRuntimeHelpersObjectSpread({}, baseStyleProps, {
  backgroundColor: ['backgroundColor', $a739655128e17f93c5a267e67462b806$var$backgroundColorValue],
  borderWidth: ['borderWidth', $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderStartWidth: [$a739655128e17f93c5a267e67462b806$var$rtl('borderLeftWidth', 'borderRightWidth'), $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderEndWidth: [$a739655128e17f93c5a267e67462b806$var$rtl('borderRightWidth', 'borderLeftWidth'), $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderLeftWidth: ['borderLeftWidth', $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderRightWidth: ['borderRightWidth', $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderTopWidth: ['borderTopWidth', $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderBottomWidth: ['borderBottomWidth', $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderXWidth: [['borderLeftWidth', 'borderRightWidth'], $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderYWidth: [['borderTopWidth', 'borderBottomWidth'], $a739655128e17f93c5a267e67462b806$var$borderSizeValue],
  borderColor: ['borderColor', $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderStartColor: [$a739655128e17f93c5a267e67462b806$var$rtl('borderLeftColor', 'borderRightColor'), $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderEndColor: [$a739655128e17f93c5a267e67462b806$var$rtl('borderRightColor', 'borderLeftColor'), $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderLeftColor: ['borderLeftColor', $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderRightColor: ['borderRightColor', $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderTopColor: ['borderTopColor', $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderBottomColor: ['borderBottomColor', $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderXColor: [['borderLeftColor', 'borderRightColor'], $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderYColor: [['borderTopColor', 'borderBottomColor'], $a739655128e17f93c5a267e67462b806$var$borderColorValue],
  borderRadius: ['borderRadius', $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderTopStartRadius: [$a739655128e17f93c5a267e67462b806$var$rtl('borderTopLeftRadius', 'borderTopRightRadius'), $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderTopEndRadius: [$a739655128e17f93c5a267e67462b806$var$rtl('borderTopRightRadius', 'borderTopLeftRadius'), $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderBottomStartRadius: [$a739655128e17f93c5a267e67462b806$var$rtl('borderBottomLeftRadius', 'borderBottomRightRadius'), $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderBottomEndRadius: [$a739655128e17f93c5a267e67462b806$var$rtl('borderBottomRightRadius', 'borderBottomLeftRadius'), $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderTopLeftRadius: ['borderTopLeftRadius', $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderTopRightRadius: ['borderTopRightRadius', $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderBottomLeftRadius: ['borderBottomLeftRadius', $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  borderBottomRightRadius: ['borderBottomRightRadius', $a739655128e17f93c5a267e67462b806$var$borderRadiusValue],
  padding: ['padding', dimensionValue],
  paddingStart: [$a739655128e17f93c5a267e67462b806$var$rtl('paddingLeft', 'paddingRight'), dimensionValue],
  paddingEnd: [$a739655128e17f93c5a267e67462b806$var$rtl('paddingRight', 'paddingLeft'), dimensionValue],
  paddingLeft: ['paddingLeft', dimensionValue],
  paddingRight: ['paddingRight', dimensionValue],
  paddingTop: ['paddingTop', dimensionValue],
  paddingBottom: ['paddingBottom', dimensionValue],
  paddingX: [['paddingLeft', 'paddingRight'], dimensionValue],
  paddingY: [['paddingTop', 'paddingBottom'], dimensionValue]
});

exports.viewStyleProps = viewStyleProps;
var $a739655128e17f93c5a267e67462b806$var$borderStyleProps = {
  borderWidth: 'borderStyle',
  borderLeftWidth: 'borderLeftStyle',
  borderRightWidth: 'borderRightStyle',
  borderTopWidth: 'borderTopStyle',
  borderBottomWidth: 'borderBottomStyle'
};

function $a739655128e17f93c5a267e67462b806$var$rtl(ltr, rtl) {
  return function (direction) {
    return direction === 'rtl' ? rtl : ltr;
  };
}

function dimensionValue(value) {
  if (typeof value === 'number') {
    return value + 'px';
  }

  if (/(%|px|em|rem|vw)$/.test(value)) {
    return value;
  }

  return "var(--spectrum-global-dimension-".concat(value, ", var(--spectrum-alias-").concat(value, "))");
}

exports.dimensionValue = dimensionValue;

function $a739655128e17f93c5a267e67462b806$var$colorValue(value) {
  var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'default';
  return "var(--spectrum-global-color-".concat(value, ", var(--spectrum-semantic-").concat(value, "-color-").concat(type, "))");
}

function $a739655128e17f93c5a267e67462b806$var$backgroundColorValue(value) {
  return "var(--spectrum-alias-background-color-".concat(value, ", ").concat($a739655128e17f93c5a267e67462b806$var$colorValue(value, 'background'), ")");
}

function $a739655128e17f93c5a267e67462b806$var$borderColorValue(value) {
  if (value === 'default') {
    return 'var(--spectrum-alias-border-color)';
  }

  return "var(--spectrum-alias-border-color-".concat(value, ", ").concat($a739655128e17f93c5a267e67462b806$var$colorValue(value, 'border'), ")");
}

function $a739655128e17f93c5a267e67462b806$var$borderSizeValue(value) {
  return "var(--spectrum-alias-border-size-".concat(value, ")");
}

function $a739655128e17f93c5a267e67462b806$var$borderRadiusValue(value) {
  return "var(--spectrum-alias-border-radius-".concat(value, ")");
}

function $a739655128e17f93c5a267e67462b806$var$hiddenValue(value) {
  return value ? 'none' : undefined;
}

function $a739655128e17f93c5a267e67462b806$var$anyValue(value) {
  return value;
}

function convertStyleProps(props, handlers, direction) {
  var style = {};

  for (var _key in props) {
    var styleProp = handlers[_key];

    if (!styleProp || props[_key] == null) {
      continue;
    }

    var _styleProp = _babelRuntimeHelpersSlicedToArray(styleProp, 2),
        name = _styleProp[0],
        convert = _styleProp[1];

    if (typeof name === 'function') {
      name = name(direction);
    }

    var _value = convert(props[_key]);

    if (Array.isArray(name)) {
      var _iterator = $a739655128e17f93c5a267e67462b806$var$_createForOfIteratorHelper(name),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var k = _step.value;
          style[k] = _value;
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    } else {
      style[name] = _value;
    }
  }

  for (var prop in $a739655128e17f93c5a267e67462b806$var$borderStyleProps) {
    if (style[prop]) {
      style[$a739655128e17f93c5a267e67462b806$var$borderStyleProps[prop]] = 'solid';
      style.boxSizing = 'border-box';
    }
  }

  return style;
}

exports.convertStyleProps = convertStyleProps;

function useStyleProps(props) {
  var handlers = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : baseStyleProps;

  var UNSAFE_className = props.UNSAFE_className,
      UNSAFE_style = props.UNSAFE_style,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(props, ["UNSAFE_className", "UNSAFE_style"]);

  var _useLocale = useLocale(),
      direction = _useLocale.direction;

  var styles = convertStyleProps(props, handlers, direction);

  var style = _babelRuntimeHelpersObjectSpread({}, UNSAFE_style, {}, styles); // @ts-ignore


  if (otherProps.className) {
    console.warn('The className prop is unsafe and is unsupported in React Spectrum v3. ' + 'Please use style props with Spectrum variables, or UNSAFE_className if you absolutely must to something custom. ' + 'Note that this may break in future versions due to DOM structure changes.');
  } // @ts-ignore


  if (otherProps.style) {
    console.warn('The style prop is unsafe and is unsupported in React Spectrum v3. ' + 'Please use style props with Spectrum variables, or UNSAFE_style if you absolutely must to something custom. ' + 'Note that this may break in future versions due to DOM structure changes.');
  }

  var styleProps = {
    style: style,
    className: UNSAFE_className
  };

  if (props.isHidden) {
    styleProps.hidden = true;
  }

  return {
    styleProps: styleProps
  };
}

exports.useStyleProps = useStyleProps;

function passthroughStyle(value) {
  return value;
}

exports.passthroughStyle = passthroughStyle;

var $a739655128e17f93c5a267e67462b806$var$boxAlignmentStyleProps = _babelRuntimeHelpersObjectSpread({
  justifyItems: ['justify-items', passthroughStyle],
  justifyContent: ['justify-content', passthroughStyle],
  alignItems: ['align-items', passthroughStyle],
  alignContent: ['align-content', passthroughStyle]
}, baseStyleProps);

var flexStyleProps = _babelRuntimeHelpersObjectSpread({
  flexDirection: ['flex-direction', passthroughStyle],
  flexWrap: ['flex-wrap', passthroughStyle],
  flexFlow: ['flex-flow', passthroughStyle]
}, $a739655128e17f93c5a267e67462b806$var$boxAlignmentStyleProps);

exports.flexStyleProps = flexStyleProps;

var gridStyleProps = _babelRuntimeHelpersObjectSpread({
  grid: ['grid', passthroughStyle],
  gridArea: ['grid-area', passthroughStyle],
  gridAutoColumns: ['grid-auto-columns', passthroughStyle],
  gridAutoFlow: ['grid-auto-flow', passthroughStyle],
  gridAutoRows: ['grid-auto-rows', passthroughStyle],
  gridColumn: ['grid-column', passthroughStyle],
  gridColumnEnd: ['grid-column-end', passthroughStyle],
  gridColumnStart: ['grid-column-start', passthroughStyle],
  gridRow: ['grid-row', passthroughStyle],
  gridRowEnd: ['grid-row-end', passthroughStyle],
  gridRowStart: ['grid-row-start', passthroughStyle],
  gridTemplate: ['grid-template', passthroughStyle],
  gridTemplateAreas: ['grid-template-areas', passthroughStyle],
  gridTemplateColumns: ['grid-template-columns', passthroughStyle],
  gridTemplateRows: ['grid-template-rows', passthroughStyle],
  rowGap: ['row-gap', dimensionValue],
  columnGap: ['row-gap', dimensionValue]
}, $a739655128e17f93c5a267e67462b806$var$boxAlignmentStyleProps);

exports.gridStyleProps = gridStyleProps;

var $dc2d7fbd7cc7a47559a70a7cb00af3$var$SlotContext = _react.createContext(null);

function useSlotProps(props, defaultSlot) {
  var slot = props.slot || defaultSlot;

  var _ref = useContext($dc2d7fbd7cc7a47559a70a7cb00af3$var$SlotContext) || {},
      _ref$slot = _ref[slot],
      slotProps = _ref$slot === void 0 ? {} : _ref$slot;

  return mergeProps(slotProps, props);
}

exports.useSlotProps = useSlotProps;

function cssModuleToSlots(cssModule) {
  return Object.keys(cssModule).reduce(function (acc, slot) {
    acc[slot] = {
      UNSAFE_className: cssModule[slot]
    };
    return acc;
  }, {});
}

exports.cssModuleToSlots = cssModuleToSlots;

function SlotProvider(props) {
  var parentSlots = useContext($dc2d7fbd7cc7a47559a70a7cb00af3$var$SlotContext) || {};
  var _props$slots = props.slots,
      slots = _props$slots === void 0 ? {} : _props$slots,
      children = props.children; // Merge props for each slot from parent context and props

  var value = useMemo(function () {
    return Object.keys(parentSlots).concat(Object.keys(slots)).reduce(function (o, p) {
      return _babelRuntimeHelpersObjectSpread({}, o, _babelRuntimeHelpersDefineProperty({}, p, mergeProps(parentSlots[p] || {}, slots[p] || {})));
    }, {});
  }, [parentSlots, slots]);
  return (/*#__PURE__*/_react.createElement($dc2d7fbd7cc7a47559a70a7cb00af3$var$SlotContext.Provider, {
      value: value
    }, children)
  );
}

exports.SlotProvider = SlotProvider;

function ClearSlots(props) {
  var children = props.children,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(props, ["children"]);

  var content = children;

  if (_react.Children.toArray(children).length <= 1) {
    if (typeof children === 'function') {
      // need to know if the node is a string or something else that react can render that doesn't get props
      content = _react.cloneElement(_react.Children.only(children), otherProps);
    }
  }

  return (/*#__PURE__*/_react.createElement($dc2d7fbd7cc7a47559a70a7cb00af3$var$SlotContext.Provider, {
      value: {}
    }, content)
  );
}

exports.ClearSlots = ClearSlots;

function useHasChild(query, ref) {
  var _useState = useState(true),
      _useState2 = _babelRuntimeHelpersSlicedToArray(_useState, 2),
      hasChild = _useState2[0],
      setHasChild = _useState2[1];

  useLayoutEffect(function () {
    setHasChild(!!(ref.current && ref.current.querySelector(query)));
  }, [setHasChild, query, ref]);
  return hasChild;
}

exports.useHasChild = useHasChild;
