// @react-spectrum/utils/dist/module.js
import _classnames from "classnames";
import _react, { useEffect, useState, useImperativeHandle, useRef, useContext, useMemo, useLayoutEffect } from "react";
import _babelRuntimeHelpersEsmObjectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _babelRuntimeHelpersEsmObjectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { useLocale } from "@react-aria/i18n";
import { mergeProps } from "@react-aria/utils";
export let shouldKeepSpectrumClassNames = false;
export function keepSpectrumClassNames() {
  shouldKeepSpectrumClassNames = true;
  console.warn('Legacy spectrum-prefixed class names enabled for backward compatibility. ' + 'We recommend replacing instances of CSS overrides targeting spectrum selectors ' + 'in your app with custom class names of your own, and disabling this flag.');
}
export function classNames(cssModule, ...values) {
  let classes = [];

  for (let value of values) {
    if (typeof value === 'object' && value) {
      let mapped = {};

      for (let key in value) {
        if (cssModule[key]) {
          mapped[cssModule[key]] = value[key];
        }

        if (shouldKeepSpectrumClassNames || !cssModule[key]) {
          mapped[key] = value[key];
        }
      }

      classes.push(mapped);
    } else if (typeof value === 'string') {
      if (cssModule[value]) {
        classes.push(cssModule[value]);
      }

      if (shouldKeepSpectrumClassNames || !cssModule[value]) {
        classes.push(value);
      }
    } else {
      classes.push(value);
    }
  }

  return _classnames(...classes);
}
export function getWrappedElement(children) {
  let element;

  if (typeof children === 'string') {
    element = /*#__PURE__*/_react.createElement("span", null, children);
  } else {
    element = _react.Children.only(children);
  }

  return element;
}

/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
// IMPORTANT: this needs to be synced with the TypeScript definition of DOMProps
const $fb89d1caeac8c7767eab7a1048978$var$DOMPropNames = {
  id: 1,
  tabIndex: 1,
  role: 1,
  'aria-label': 1,
  'aria-labelledby': 1,
  'aria-describedby': 1,
  'aria-controls': 1,
  'aria-owns': 1,
  'aria-hidden': 1
}; // Sync with TypeScript definition of TextInputDOMProps

export const TextInputDOMPropNames = {
  autoComplete: 1,
  maxLength: 1,
  minLength: 1,
  name: 1,
  pattern: 1,
  placeholder: 1,
  type: 1,
  inputMode: 1,
  // Clipboard events
  onCopy: 1,
  onCut: 1,
  onPaste: 1,
  // Composition events
  onCompositionEnd: 1,
  onCompositionStart: 1,
  onCompositionUpdate: 1,
  // Selection events
  onSelect: 1,
  // Input events
  onBeforeInput: 1,
  onInput: 1
}; // Based on DOMAttributes from react types

export const DOMEventPropNames = {
  // Focus events
  onFocus: 1,
  onBlur: 1,
  // Keyboard events
  onKeyDown: 1,
  onKeyPress: 1,
  onKeyUp: 1,
  // Mouse events
  onAuxClick: 1,
  onClick: 1,
  onContextMenu: 1,
  onDoubleClick: 1,
  onDrag: 1,
  onDragEnd: 1,
  onDragEnter: 1,
  onDragExit: 1,
  onDragLeave: 1,
  onDragOver: 1,
  onDragStart: 1,
  onDrop: 1,
  onMouseDown: 1,
  onMouseEnter: 1,
  onMouseLeave: 1,
  onMouseMove: 1,
  onMouseOut: 1,
  onMouseOver: 1,
  onMouseUp: 1,
  // Touch events
  onTouchCancel: 1,
  onTouchEnd: 1,
  onTouchMove: 1,
  onTouchStart: 1,
  // Pointer events
  onPointerDown: 1,
  onPointerMove: 1,
  onPointerUp: 1,
  onPointerCancel: 1,
  onPointerEnter: 1,
  onPointerLeave: 1,
  onPointerOver: 1,
  onPointerOut: 1,
  onGotPointerCapture: 1,
  onLostPointerCapture: 1,
  // UI events
  onScroll: 1,
  // Wheel events
  onWheel: 1
};
/**
 * Checking for data-* props
 */

const $fb89d1caeac8c7767eab7a1048978$var$propRe = /^(data-.*)$/; // Filters out all props that aren't valid DOM props or are user defined via override prop obj.

export function filterDOMProps(props, override = {}) {
  const filterProps = {};

  const propFilter = _babelRuntimeHelpersEsmObjectSpread({}, $fb89d1caeac8c7767eab7a1048978$var$DOMPropNames, {}, override);

  for (const prop in props) {
    if (Object.prototype.hasOwnProperty.call(props, prop) && (propFilter[prop] || $fb89d1caeac8c7767eab7a1048978$var$propRe.test(prop))) {
      filterProps[prop] = props[prop];
    }
  }

  return filterProps;
}
export function useMediaQuery(query) {
  let supportsMatchMedia = typeof window !== 'undefined' && typeof window.matchMedia === 'function';
  let [matches, setMatches] = useState(() => supportsMatchMedia ? window.matchMedia(query).matches : false);
  useEffect(() => {
    if (!supportsMatchMedia) {
      return;
    }

    let mq = window.matchMedia(query);

    let onChange = evt => {
      setMatches(evt.matches);
    };

    mq.addListener(onChange);
    return () => {
      mq.removeListener(onChange);
    };
  }, [supportsMatchMedia, query]);
  return matches;
}
export function createDOMRef(ref) {
  return {
    UNSAFE_getDOMNode() {
      return ref.current;
    }

  };
}
export function createFocusableRef(domRef, focusableRef = domRef) {
  return _babelRuntimeHelpersEsmObjectSpread({}, createDOMRef(domRef), {
    focus() {
      if (focusableRef.current) {
        focusableRef.current.focus();
      }
    }

  });
}
export function useDOMRef(ref) {
  let domRef = useRef(null);
  useImperativeHandle(ref, () => createDOMRef(domRef));
  return domRef;
}
export function useFocusableRef(ref, focusableRef) {
  let domRef = useRef(null);
  useImperativeHandle(ref, () => createFocusableRef(domRef, focusableRef));
  return domRef;
}
export function unwrapDOMRef(ref) {
  return {
    get current() {
      return ref.current && ref.current.UNSAFE_getDOMNode();
    }

  };
}
export const baseStyleProps = {
  margin: ['margin', dimensionValue],
  marginStart: [$a2652d3ec5e13f7201952092971eebf$var$rtl('marginLeft', 'marginRight'), dimensionValue],
  marginEnd: [$a2652d3ec5e13f7201952092971eebf$var$rtl('marginRight', 'marginLeft'), dimensionValue],
  marginLeft: ['marginLeft', dimensionValue],
  marginRight: ['marginRight', dimensionValue],
  marginTop: ['marginTop', dimensionValue],
  marginBottom: ['marginBottom', dimensionValue],
  marginX: [['marginLeft', 'marginRight'], dimensionValue],
  marginY: [['marginTop', 'marginBottom'], dimensionValue],
  width: ['width', dimensionValue],
  height: ['height', dimensionValue],
  minWidth: ['minWidth', dimensionValue],
  minHeight: ['minHeight', dimensionValue],
  maxWidth: ['maxWidth', dimensionValue],
  maxHeight: ['maxHeight', dimensionValue],
  isHidden: ['display', $a2652d3ec5e13f7201952092971eebf$var$hiddenValue],
  alignSelf: ['alignSelf', passthroughStyle],
  justifySelf: ['justifySelf', passthroughStyle],
  position: ['position', $a2652d3ec5e13f7201952092971eebf$var$anyValue],
  zIndex: ['zIndex', $a2652d3ec5e13f7201952092971eebf$var$anyValue],
  top: ['top', dimensionValue],
  bottom: ['bottom', dimensionValue],
  start: [$a2652d3ec5e13f7201952092971eebf$var$rtl('left', 'right'), dimensionValue],
  end: [$a2652d3ec5e13f7201952092971eebf$var$rtl('right', 'left'), dimensionValue],
  left: ['left', dimensionValue],
  right: ['right', dimensionValue],
  flex: ['flex', passthroughStyle],
  flexGrow: ['flexGrow', passthroughStyle],
  flexShrink: ['flexShrink', passthroughStyle],
  flexBasis: ['flexBasis', passthroughStyle]
};
export const viewStyleProps = _babelRuntimeHelpersEsmObjectSpread({}, baseStyleProps, {
  backgroundColor: ['backgroundColor', $a2652d3ec5e13f7201952092971eebf$var$backgroundColorValue],
  borderWidth: ['borderWidth', $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderStartWidth: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderLeftWidth', 'borderRightWidth'), $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderEndWidth: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderRightWidth', 'borderLeftWidth'), $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderLeftWidth: ['borderLeftWidth', $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderRightWidth: ['borderRightWidth', $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderTopWidth: ['borderTopWidth', $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderBottomWidth: ['borderBottomWidth', $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderXWidth: [['borderLeftWidth', 'borderRightWidth'], $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderYWidth: [['borderTopWidth', 'borderBottomWidth'], $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue],
  borderColor: ['borderColor', $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderStartColor: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderLeftColor', 'borderRightColor'), $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderEndColor: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderRightColor', 'borderLeftColor'), $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderLeftColor: ['borderLeftColor', $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderRightColor: ['borderRightColor', $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderTopColor: ['borderTopColor', $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderBottomColor: ['borderBottomColor', $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderXColor: [['borderLeftColor', 'borderRightColor'], $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderYColor: [['borderTopColor', 'borderBottomColor'], $a2652d3ec5e13f7201952092971eebf$var$borderColorValue],
  borderRadius: ['borderRadius', $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderTopStartRadius: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderTopLeftRadius', 'borderTopRightRadius'), $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderTopEndRadius: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderTopRightRadius', 'borderTopLeftRadius'), $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderBottomStartRadius: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderBottomLeftRadius', 'borderBottomRightRadius'), $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderBottomEndRadius: [$a2652d3ec5e13f7201952092971eebf$var$rtl('borderBottomRightRadius', 'borderBottomLeftRadius'), $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderTopLeftRadius: ['borderTopLeftRadius', $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderTopRightRadius: ['borderTopRightRadius', $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderBottomLeftRadius: ['borderBottomLeftRadius', $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  borderBottomRightRadius: ['borderBottomRightRadius', $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue],
  padding: ['padding', dimensionValue],
  paddingStart: [$a2652d3ec5e13f7201952092971eebf$var$rtl('paddingLeft', 'paddingRight'), dimensionValue],
  paddingEnd: [$a2652d3ec5e13f7201952092971eebf$var$rtl('paddingRight', 'paddingLeft'), dimensionValue],
  paddingLeft: ['paddingLeft', dimensionValue],
  paddingRight: ['paddingRight', dimensionValue],
  paddingTop: ['paddingTop', dimensionValue],
  paddingBottom: ['paddingBottom', dimensionValue],
  paddingX: [['paddingLeft', 'paddingRight'], dimensionValue],
  paddingY: [['paddingTop', 'paddingBottom'], dimensionValue]
});
const $a2652d3ec5e13f7201952092971eebf$var$borderStyleProps = {
  borderWidth: 'borderStyle',
  borderLeftWidth: 'borderLeftStyle',
  borderRightWidth: 'borderRightStyle',
  borderTopWidth: 'borderTopStyle',
  borderBottomWidth: 'borderBottomStyle'
};

function $a2652d3ec5e13f7201952092971eebf$var$rtl(ltr, rtl) {
  return direction => direction === 'rtl' ? rtl : ltr;
}

export function dimensionValue(value) {
  if (typeof value === 'number') {
    return value + 'px';
  }

  if (/(%|px|em|rem|vw)$/.test(value)) {
    return value;
  }

  return "var(--spectrum-global-dimension-".concat(value, ", var(--spectrum-alias-").concat(value, "))");
}

function $a2652d3ec5e13f7201952092971eebf$var$colorValue(value, type = 'default') {
  return "var(--spectrum-global-color-".concat(value, ", var(--spectrum-semantic-").concat(value, "-color-").concat(type, "))");
}

function $a2652d3ec5e13f7201952092971eebf$var$backgroundColorValue(value) {
  return "var(--spectrum-alias-background-color-".concat(value, ", ").concat($a2652d3ec5e13f7201952092971eebf$var$colorValue(value, 'background'), ")");
}

function $a2652d3ec5e13f7201952092971eebf$var$borderColorValue(value) {
  if (value === 'default') {
    return 'var(--spectrum-alias-border-color)';
  }

  return "var(--spectrum-alias-border-color-".concat(value, ", ").concat($a2652d3ec5e13f7201952092971eebf$var$colorValue(value, 'border'), ")");
}

function $a2652d3ec5e13f7201952092971eebf$var$borderSizeValue(value) {
  return "var(--spectrum-alias-border-size-".concat(value, ")");
}

function $a2652d3ec5e13f7201952092971eebf$var$borderRadiusValue(value) {
  return "var(--spectrum-alias-border-radius-".concat(value, ")");
}

function $a2652d3ec5e13f7201952092971eebf$var$hiddenValue(value) {
  return value ? 'none' : undefined;
}

function $a2652d3ec5e13f7201952092971eebf$var$anyValue(value) {
  return value;
}

export function convertStyleProps(props, handlers, direction) {
  let style = {};

  for (let key in props) {
    let styleProp = handlers[key];

    if (!styleProp || props[key] == null) {
      continue;
    }

    let [name, convert] = styleProp;

    if (typeof name === 'function') {
      name = name(direction);
    }

    let value = convert(props[key]);

    if (Array.isArray(name)) {
      for (let k of name) {
        style[k] = value;
      }
    } else {
      style[name] = value;
    }
  }

  for (let prop in $a2652d3ec5e13f7201952092971eebf$var$borderStyleProps) {
    if (style[prop]) {
      style[$a2652d3ec5e13f7201952092971eebf$var$borderStyleProps[prop]] = 'solid';
      style.boxSizing = 'border-box';
    }
  }

  return style;
}
export function useStyleProps(props, handlers = baseStyleProps) {
  let {
    UNSAFE_className,
    UNSAFE_style
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["UNSAFE_className", "UNSAFE_style"]);

  let {
    direction
  } = useLocale();
  let styles = convertStyleProps(props, handlers, direction);

  let style = _babelRuntimeHelpersEsmObjectSpread({}, UNSAFE_style, {}, styles); // @ts-ignore


  if (otherProps.className) {
    console.warn('The className prop is unsafe and is unsupported in React Spectrum v3. ' + 'Please use style props with Spectrum variables, or UNSAFE_className if you absolutely must to something custom. ' + 'Note that this may break in future versions due to DOM structure changes.');
  } // @ts-ignore


  if (otherProps.style) {
    console.warn('The style prop is unsafe and is unsupported in React Spectrum v3. ' + 'Please use style props with Spectrum variables, or UNSAFE_style if you absolutely must to something custom. ' + 'Note that this may break in future versions due to DOM structure changes.');
  }

  let styleProps = {
    style,
    className: UNSAFE_className
  };

  if (props.isHidden) {
    styleProps.hidden = true;
  }

  return {
    styleProps
  };
}
export function passthroughStyle(value) {
  return value;
}

const $a2652d3ec5e13f7201952092971eebf$var$boxAlignmentStyleProps = _babelRuntimeHelpersEsmObjectSpread({
  justifyItems: ['justify-items', passthroughStyle],
  justifyContent: ['justify-content', passthroughStyle],
  alignItems: ['align-items', passthroughStyle],
  alignContent: ['align-content', passthroughStyle]
}, baseStyleProps);

export const flexStyleProps = _babelRuntimeHelpersEsmObjectSpread({
  flexDirection: ['flex-direction', passthroughStyle],
  flexWrap: ['flex-wrap', passthroughStyle],
  flexFlow: ['flex-flow', passthroughStyle]
}, $a2652d3ec5e13f7201952092971eebf$var$boxAlignmentStyleProps);
export const gridStyleProps = _babelRuntimeHelpersEsmObjectSpread({
  grid: ['grid', passthroughStyle],
  gridArea: ['grid-area', passthroughStyle],
  gridAutoColumns: ['grid-auto-columns', passthroughStyle],
  gridAutoFlow: ['grid-auto-flow', passthroughStyle],
  gridAutoRows: ['grid-auto-rows', passthroughStyle],
  gridColumn: ['grid-column', passthroughStyle],
  gridColumnEnd: ['grid-column-end', passthroughStyle],
  gridColumnStart: ['grid-column-start', passthroughStyle],
  gridRow: ['grid-row', passthroughStyle],
  gridRowEnd: ['grid-row-end', passthroughStyle],
  gridRowStart: ['grid-row-start', passthroughStyle],
  gridTemplate: ['grid-template', passthroughStyle],
  gridTemplateAreas: ['grid-template-areas', passthroughStyle],
  gridTemplateColumns: ['grid-template-columns', passthroughStyle],
  gridTemplateRows: ['grid-template-rows', passthroughStyle],
  rowGap: ['row-gap', dimensionValue],
  columnGap: ['row-gap', dimensionValue]
}, $a2652d3ec5e13f7201952092971eebf$var$boxAlignmentStyleProps);

let $ac9a73b963439a7b338bb32dc8132d8$var$SlotContext = _react.createContext(null);

export function useSlotProps(props, defaultSlot) {
  let slot = props.slot || defaultSlot;
  let {
    [slot]: slotProps = {}
  } = useContext($ac9a73b963439a7b338bb32dc8132d8$var$SlotContext) || {};
  return mergeProps(slotProps, props);
}
export function cssModuleToSlots(cssModule) {
  return Object.keys(cssModule).reduce((acc, slot) => {
    acc[slot] = {
      UNSAFE_className: cssModule[slot]
    };
    return acc;
  }, {});
}
export function SlotProvider(props) {
  let parentSlots = useContext($ac9a73b963439a7b338bb32dc8132d8$var$SlotContext) || {};
  let {
    slots = {},
    children
  } = props; // Merge props for each slot from parent context and props

  let value = useMemo(() => Object.keys(parentSlots).concat(Object.keys(slots)).reduce((o, p) => _babelRuntimeHelpersEsmObjectSpread({}, o, {
    [p]: mergeProps(parentSlots[p] || {}, slots[p] || {})
  }), {}), [parentSlots, slots]);
  return (/*#__PURE__*/_react.createElement($ac9a73b963439a7b338bb32dc8132d8$var$SlotContext.Provider, {
      value: value
    }, children)
  );
}
export function ClearSlots(props) {
  let {
    children
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["children"]);

  let content = children;

  if (_react.Children.toArray(children).length <= 1) {
    if (typeof children === 'function') {
      // need to know if the node is a string or something else that react can render that doesn't get props
      content = _react.cloneElement(_react.Children.only(children), otherProps);
    }
  }

  return (/*#__PURE__*/_react.createElement($ac9a73b963439a7b338bb32dc8132d8$var$SlotContext.Provider, {
      value: {}
    }, content)
  );
}
export function useHasChild(query, ref) {
  let [hasChild, setHasChild] = useState(true);
  useLayoutEffect(() => {
    setHasChild(!!(ref.current && ref.current.querySelector(query)));
  }, [setHasChild, query, ref]);
  return hasChild;
}
