// @react-spectrum/typography/dist/main.js
var _babelRuntimeHelpersExtends = $parcel$interopDefault(require("@babel/runtime/helpers/extends"));

var _babelRuntimeHelpersObjectWithoutProperties = $parcel$interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _temp = require("@react-spectrum/utils");

var filterDOMProps = _temp.filterDOMProps;
var useDOMRef = _temp.useDOMRef;
var useSlotProps = _temp.useSlotProps;
var useStyleProps = _temp.useStyleProps;

var _react2 = require("react");

var _react = $parcel$interopDefault(_react2);

var forwardRef = _react2.forwardRef;

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

function $afd0a7941697f5db81e3b4387b8acfab$var$Text(props, ref) {
  props = useSlotProps(props, 'text');

  var _props = props,
      children = _props.children,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(_props, ["children"]);

  var _useStyleProps = useStyleProps(otherProps),
      styleProps = _useStyleProps.styleProps;

  var domRef = useDOMRef(ref);
  return (/*#__PURE__*/_react.createElement("span", _babelRuntimeHelpersExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef
    }), children)
  );
}

var Text = forwardRef($afd0a7941697f5db81e3b4387b8acfab$var$Text);
exports.Text = Text;

function $af1f3bc3e1140ced6146e2069$var$Heading(props, ref) {
  props = useSlotProps(props, 'heading');

  var _props = props,
      children = _props.children,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(_props, ["children"]);

  var _useStyleProps = useStyleProps(otherProps),
      styleProps = _useStyleProps.styleProps;

  var domRef = useDOMRef(ref);
  return (/*#__PURE__*/_react.createElement("h1", _babelRuntimeHelpersExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef
    }), children)
  );
}

var Heading = forwardRef($af1f3bc3e1140ced6146e2069$var$Heading);
exports.Heading = Heading;

function $c76dd28494f0aeb782562911e00dd988$var$Keyboard(props, ref) {
  props = useSlotProps(props, 'keyboard');

  var _props = props,
      children = _props.children,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(_props, ["children"]);

  var _useStyleProps = useStyleProps(otherProps),
      styleProps = _useStyleProps.styleProps;

  var domRef = useDOMRef(ref);
  return (/*#__PURE__*/_react.createElement("kbd", _babelRuntimeHelpersExtends({}, filterDOMProps(otherProps), styleProps, {
      dir: "ltr",
      ref: domRef
    }), children)
  );
}

var Keyboard = forwardRef($c76dd28494f0aeb782562911e00dd988$var$Keyboard);
exports.Keyboard = Keyboard;
