// @react-spectrum/typography/dist/module.js
import _babelRuntimeHelpersEsmExtends from "@babel/runtime/helpers/esm/extends";
import _babelRuntimeHelpersEsmObjectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { filterDOMProps, useDOMRef, useSlotProps, useStyleProps } from "@react-spectrum/utils";
import _react, { forwardRef } from "react";

function $dab6698a8604d811090b01ce1ed1cc22$var$Text(props, ref) {
  props = useSlotProps(props, 'text');

  let {
    children
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["children"]);

  let {
    styleProps
  } = useStyleProps(otherProps);
  let domRef = useDOMRef(ref);
  return (/*#__PURE__*/_react.createElement("span", _babelRuntimeHelpersEsmExtends({}, filterDOMProps(otherProps), styleProps, {
      ref: domRef
    }), children)
  );
}

export const Text = forwardRef($dab6698a8604d811090b01ce1ed1cc22$var$Text);

function $fafc33a2e1ffa1042b6e360f41db2bb$var$Heading(props, ref) {
  props = useSlotProps(props, 'heading');

  let {
    children
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["children"]);

  let {
    styleProps
  } = useStyleProps(otherProps);
  let domRef = useDOMRef(ref);
  return _react.createElement("h1", _babelRuntimeHelpersEsmExtends({}, filterDOMProps(otherProps), styleProps, {
    ref: domRef
  }), children);
}

export const Heading = forwardRef($fafc33a2e1ffa1042b6e360f41db2bb$var$Heading);

function $f839fe3a01c16b34feb0028801dd95a7$var$Keyboard(props, ref) {
  props = useSlotProps(props, 'keyboard');

  let {
    children
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutProperties(props, ["children"]);

  let {
    styleProps
  } = useStyleProps(otherProps);
  let domRef = useDOMRef(ref);
  return (/*#__PURE__*/_react.createElement("kbd", _babelRuntimeHelpersEsmExtends({}, filterDOMProps(otherProps), styleProps, {
      dir: "ltr",
      ref: domRef
    }), children)
  );
}

export const Keyboard = forwardRef($f839fe3a01c16b34feb0028801dd95a7$var$Keyboard);
