// @react-spectrum/checkbox/dist/main.js
require("./main.css");

var _babelRuntimeHelpersExtends = $parcel$interopDefault(require("@babel/runtime/helpers/extends"));

var _babelRuntimeHelpersObjectWithoutProperties = $parcel$interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _spectrumIconsUiCheckmarkSmall = $parcel$interopDefault(require("@spectrum-icons/ui/CheckmarkSmall"));

var _temp = require("@react-spectrum/utils");

var classNames = _temp.classNames;
var filterDOMProps = _temp.filterDOMProps;
var useFocusableRef = _temp.useFocusableRef;
var useStyleProps = _temp.useStyleProps;

var _spectrumIconsUiDashSmall = $parcel$interopDefault(require("@spectrum-icons/ui/DashSmall"));

var FocusRing = require("@react-aria/focus").FocusRing;

var _react2 = require("react");

var _react = $parcel$interopDefault(_react2);

var forwardRef = _react2.forwardRef;
var useRef = _react2.useRef;

var useCheckbox = require("@react-aria/checkbox").useCheckbox;

var useProviderProps = require("@react-spectrum/provider").useProviderProps;

var useToggleState = require("@react-stately/toggle").useToggleState;

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/components/checkbox/vars.css
var $e062db7c5c4d807e6f44b0953977b0d2$exports = {};
$e062db7c5c4d807e6f44b0953977b0d2$exports = {
  "spectrum-Checkbox": "_spectrum-Checkbox_21b0f",
  "spectrum-Checkbox-input": "_spectrum-Checkbox-input_21b0f",
  "spectrum-Checkbox-box": "_spectrum-Checkbox-box_21b0f",
  "spectrum-Checkbox-checkmark": "_spectrum-Checkbox-checkmark_21b0f",
  "is-indeterminate": "_is-indeterminate_21b0f",
  "spectrum-Checkbox-partialCheckmark": "_spectrum-Checkbox-partialCheckmark_21b0f",
  "spectrum-Checkbox-label": "_spectrum-Checkbox-label_21b0f",
  "focus-ring": "_focus-ring_21b0f",
  "spectrum-Checkbox--quiet": "_spectrum-Checkbox--quiet_21b0f",
  "is-invalid": "_is-invalid_21b0f"
};
var $e062db7c5c4d807e6f44b0953977b0d2$$interop$default = $parcel$interopDefault($e062db7c5c4d807e6f44b0953977b0d2$exports);

function $aa36424a9b6d1ce727213c3c8392e35$var$Checkbox(props, ref) {
  props = useProviderProps(props);

  var _props = props,
      _props$isIndeterminat = _props.isIndeterminate,
      isIndeterminate = _props$isIndeterminat === void 0 ? false : _props$isIndeterminat,
      _props$isEmphasized = _props.isEmphasized,
      isEmphasized = _props$isEmphasized === void 0 ? false : _props$isEmphasized,
      _props$isDisabled = _props.isDisabled,
      isDisabled = _props$isDisabled === void 0 ? false : _props$isDisabled,
      autoFocus = _props.autoFocus,
      children = _props.children,
      otherProps = _babelRuntimeHelpersObjectWithoutProperties(_props, ["isIndeterminate", "isEmphasized", "isDisabled", "autoFocus", "children"]);

  var _useStyleProps = useStyleProps(otherProps),
      styleProps = _useStyleProps.styleProps;

  var inputRef = useRef(null);
  var domRef = useFocusableRef(ref, inputRef);
  var state = useToggleState(props);

  var _useCheckbox = useCheckbox(props, state, inputRef),
      inputProps = _useCheckbox.inputProps;

  var markIcon = isIndeterminate ? /*#__PURE__*/_react.createElement(_spectrumIconsUiDashSmall, {
    UNSAFE_className: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'spectrum-Checkbox-partialCheckmark')
  }) : /*#__PURE__*/_react.createElement(_spectrumIconsUiCheckmarkSmall, {
    UNSAFE_className: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'spectrum-Checkbox-checkmark')
  });
  return (/*#__PURE__*/_react.createElement("label", _babelRuntimeHelpersExtends({}, filterDOMProps(otherProps, {
      'aria-label': false
    }), styleProps, {
      ref: domRef,
      className: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'spectrum-Checkbox', {
        'is-indeterminate': isIndeterminate,
        'spectrum-Checkbox--quiet': !isEmphasized,
        'is-invalid': inputProps['aria-invalid'],
        'is-disabled': isDisabled
      }, styleProps.className)
    }), /*#__PURE__*/_react.createElement(FocusRing, {
      focusRingClass: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'focus-ring'),
      autoFocus: autoFocus
    }, /*#__PURE__*/_react.createElement("input", _babelRuntimeHelpersExtends({}, inputProps, {
      ref: inputRef,
      className: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'spectrum-Checkbox-input')
    }))), /*#__PURE__*/_react.createElement("span", {
      className: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'spectrum-Checkbox-box')
    }, markIcon), children && /*#__PURE__*/_react.createElement("span", {
      className: classNames($e062db7c5c4d807e6f44b0953977b0d2$$interop$default, 'spectrum-Checkbox-label')
    }, children))
  );
}
/**
 * Checkboxes allow users to select multiple items from a list of individual items,
 * or to mark one individual item as selected.
 */


var Checkbox = forwardRef($aa36424a9b6d1ce727213c3c8392e35$var$Checkbox);
exports.Checkbox = Checkbox;
