// @react-spectrum/theme-default/dist/main.js
require("./main.css");

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-dark.css
var $f0819d428e8f03bf0c1fcd54bf$exports = {};
$f0819d428e8f03bf0c1fcd54bf$exports = {
  "spectrum--dark": "_spectrum--dark_027c9"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-global.css
var $fe81f0edf6cd7cecfe75e0781ca5f9$exports = {};
$fe81f0edf6cd7cecfe75e0781ca5f9$exports = {
  "spectrum": "_spectrum_1cdad",
  "spectrum--medium": "_spectrum--medium_1cdad",
  "spectrum--large": "_spectrum--large_1cdad",
  "spectrum--darkest": "_spectrum--darkest_1cdad",
  "spectrum--dark": "_spectrum--dark_1cdad",
  "spectrum--light": "_spectrum--light_1cdad",
  "spectrum--lightest": "_spectrum--lightest_1cdad"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-large.css
var $eeba4c8ebe03e8344f8bbbefd13e8c$exports = {};
$eeba4c8ebe03e8344f8bbbefd13e8c$exports = {
  "spectrum--large": "_spectrum--large_c261d"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-light.css
var $fc7093def42adb3db61eb3e8399f624$exports = {};
$fc7093def42adb3db61eb3e8399f624$exports = {
  "spectrum--light": "_spectrum--light_5c408"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-medium.css
var $f7a22611fe61832566aee60e4a$exports = {};
$f7a22611fe61832566aee60e4a$exports = {
  "spectrum--medium": "_spectrum--medium_b8f4e"
};
var $fe81f0edf6cd7cecfe75e0781ca5f9$$interop$default = $parcel$interopDefault($fe81f0edf6cd7cecfe75e0781ca5f9$exports);
var $fc7093def42adb3db61eb3e8399f624$$interop$default = $parcel$interopDefault($fc7093def42adb3db61eb3e8399f624$exports);
var $f0819d428e8f03bf0c1fcd54bf$$interop$default = $parcel$interopDefault($f0819d428e8f03bf0c1fcd54bf$exports);
var $f7a22611fe61832566aee60e4a$$interop$default = $parcel$interopDefault($f7a22611fe61832566aee60e4a$exports);
var $eeba4c8ebe03e8344f8bbbefd13e8c$$interop$default = $parcel$interopDefault($eeba4c8ebe03e8344f8bbbefd13e8c$exports);
var theme = {
  global: $fe81f0edf6cd7cecfe75e0781ca5f9$$interop$default,
  light: $fc7093def42adb3db61eb3e8399f624$$interop$default,
  dark: $f0819d428e8f03bf0c1fcd54bf$$interop$default,
  medium: $f7a22611fe61832566aee60e4a$$interop$default,
  large: $eeba4c8ebe03e8344f8bbbefd13e8c$$interop$default
};
exports.theme = theme;
