// @react-spectrum/theme-default/dist/module.js
import "./main.css";

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-dark.css
var $dbf1bdb203aaef4a978eae2a76c7a8$exports = {};
$dbf1bdb203aaef4a978eae2a76c7a8$exports = {
  "spectrum--dark": "_spectrum--dark_027c9"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-global.css
var $c692dbf7671f5bf990f943a2bb56da2$exports = {};
$c692dbf7671f5bf990f943a2bb56da2$exports = {
  "spectrum": "_spectrum_1cdad",
  "spectrum--medium": "_spectrum--medium_1cdad",
  "spectrum--large": "_spectrum--large_1cdad",
  "spectrum--darkest": "_spectrum--darkest_1cdad",
  "spectrum--dark": "_spectrum--dark_1cdad",
  "spectrum--light": "_spectrum--light_1cdad",
  "spectrum--lightest": "_spectrum--lightest_1cdad"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-large.css
var $adc0e1f27956e99adbaeec13c5dbc6$exports = {};
$adc0e1f27956e99adbaeec13c5dbc6$exports = {
  "spectrum--large": "_spectrum--large_c261d"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-light.css
var $a57ea5d410727ec587a57656c44ad98$exports = {};
$a57ea5d410727ec587a57656c44ad98$exports = {
  "spectrum--light": "_spectrum--light_5c408"
};
// ASSET: /Users/govett/dev/react-spectrum/node_modules/@adobe/spectrum-css-temp/vars/spectrum-medium.css
var $e699adc2cd2d502d3a9541cb5fa294d$exports = {};
$e699adc2cd2d502d3a9541cb5fa294d$exports = {
  "spectrum--medium": "_spectrum--medium_b8f4e"
};
var $c692dbf7671f5bf990f943a2bb56da2$$interop$default = $parcel$interopDefault($c692dbf7671f5bf990f943a2bb56da2$exports);
var $a57ea5d410727ec587a57656c44ad98$$interop$default = $parcel$interopDefault($a57ea5d410727ec587a57656c44ad98$exports);
var $dbf1bdb203aaef4a978eae2a76c7a8$$interop$default = $parcel$interopDefault($dbf1bdb203aaef4a978eae2a76c7a8$exports);
var $e699adc2cd2d502d3a9541cb5fa294d$$interop$default = $parcel$interopDefault($e699adc2cd2d502d3a9541cb5fa294d$exports);
var $adc0e1f27956e99adbaeec13c5dbc6$$interop$default = $parcel$interopDefault($adc0e1f27956e99adbaeec13c5dbc6$exports);
export let theme = {
  global: $c692dbf7671f5bf990f943a2bb56da2$$interop$default,
  light: $a57ea5d410727ec587a57656c44ad98$$interop$default,
  dark: $dbf1bdb203aaef4a978eae2a76c7a8$$interop$default,
  medium: $e699adc2cd2d502d3a9541cb5fa294d$$interop$default,
  large: $adc0e1f27956e99adbaeec13c5dbc6$$interop$default
};
